// ngx-fomantic-ui/esm2015/misc/util/helpers/custom-validator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALIDATORS } from '@angular/forms';
import { forwardRef } from '@angular/core';
/**
 * @record
 */
export function ICustomValidatorHost() { }
if (false) {
    /**
     * @param {?} c
     * @return {?}
     */
    ICustomValidatorHost.prototype.validate = function (c) { };
}
/**
 * @template T
 */
export class CustomValidator {
    /**
     * @param {?} _host
     */
    constructor(_host) {
        this._host = _host;
        this.onValidatorChange = (/**
         * @return {?}
         */
        () => {
        });
    }
    /**
     * @param {?} c
     * @return {?}
     */
    validate(c) {
        return this._host.validate(c);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnValidatorChange(fn) {
        this.onValidatorChange = fn;
    }
}
if (false) {
    /** @type {?} */
    CustomValidator.prototype.onValidatorChange;
    /**
     * @type {?}
     * @private
     */
    CustomValidator.prototype._host;
}
/**
 * @record
 */
export function IValidationProvider() { }
if (false) {
    /** @type {?} */
    IValidationProvider.prototype.provide;
    /** @type {?} */
    IValidationProvider.prototype.useExisting;
    /** @type {?} */
    IValidationProvider.prototype.multi;
}
/**
 * @param {?} type
 * @return {?}
 */
export function customValidatorFactory(type) {
    return {
        provide: NG_VALIDATORS,
        useExisting: forwardRef((/**
         * @return {?}
         */
        () => type)),
        multi: true
    };
}
