// ngx-fomantic-ui/esm2015/modules/checkbox/components/checkbox.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Directive, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, ViewChild } from '@angular/core';
import { CustomValueAccessor, customValueAccessorFactory } from '../../../misc/util/internal';
export class FuiCheckbox {
    constructor() {
        this.isChecked = false;
        this.onCheckChange = new EventEmitter();
        this.onTouched = new EventEmitter();
        this.isDisabled = false;
        this.isReadonly = false;
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get checkedAttribute() {
        return this.isChecked ? '' : undefined;
    }
    /**
     * @return {?}
     */
    get isDisabledAttribute() {
        return this.isDisabled ? 'disabled' : undefined;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onMouseDown(e) {
        e.preventDefault();
    }
    /**
     * @return {?}
     */
    onClick() {
        if (!this.isDisabled && !this.isReadonly) {
            this.toggle();
            this.focusCheckbox();
        }
    }
    /**
     * @return {?}
     */
    onFocusOut() {
        this.onTouched.emit();
    }
    /**
     * @return {?}
     */
    toggle() {
        this.isChecked = !this.isChecked;
        this.onCheckChange.emit(this.isChecked);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.isChecked = value;
    }
    /**
     * @private
     * @return {?}
     */
    focusCheckbox() {
        this._checkboxElement.nativeElement.focus();
    }
}
FuiCheckbox.decorators = [
    { type: Component, args: [{
                selector: 'fui-checkbox',
                exportAs: 'fuiCheckbox',
                template: `
<input class="hidden"
       type="checkbox"
       [attr.name]="name"
       [attr.checked]="checkedAttribute"
       [attr.disabled]="isDisabledAttribute"
       [(ngModel)]="isChecked"
       #checkbox>
<label>
    <ng-content></ng-content>
</label>
`
            }] }
];
/** @nocollapse */
FuiCheckbox.ctorParameters = () => [];
FuiCheckbox.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.checkbox',] }],
    name: [{ type: Input }],
    isChecked: [{ type: HostBinding, args: ['class.checked',] }],
    onCheckChange: [{ type: Output, args: ['checkChange',] }],
    onTouched: [{ type: Output, args: ['touched',] }],
    isDisabled: [{ type: Input }],
    isReadonly: [{ type: HostBinding, args: ['class.read-only',] }, { type: Input }],
    _checkboxElement: [{ type: ViewChild, args: ['checkbox', { static: false },] }],
    onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
    onClick: [{ type: HostListener, args: ['click',] }],
    onFocusOut: [{ type: HostListener, args: ['focusout',] }]
};
if (false) {
    /** @type {?} */
    FuiCheckbox.prototype.hasClasses;
    /** @type {?} */
    FuiCheckbox.prototype.name;
    /** @type {?} */
    FuiCheckbox.prototype.isChecked;
    /** @type {?} */
    FuiCheckbox.prototype.onCheckChange;
    /** @type {?} */
    FuiCheckbox.prototype.onTouched;
    /** @type {?} */
    FuiCheckbox.prototype.isDisabled;
    /** @type {?} */
    FuiCheckbox.prototype.isReadonly;
    /**
     * @type {?}
     * @private
     */
    FuiCheckbox.prototype._checkboxElement;
}
export class FuiCheckboxValueAccessor extends CustomValueAccessor {
    /**
     * @param {?} host
     */
    constructor(host) {
        super(host);
    }
}
FuiCheckboxValueAccessor.decorators = [
    { type: Directive, args: [{
                selector: 'fui-checkbox',
                host: {
                    '(checkChange)': 'onChange($event)',
                    '(touched)': 'onTouched()'
                },
                providers: [customValueAccessorFactory(FuiCheckboxValueAccessor)]
            },] }
];
/** @nocollapse */
FuiCheckboxValueAccessor.ctorParameters = () => [
    { type: FuiCheckbox }
];
