// ngx-fomantic-ui/esm2015/modules/checkbox/components/radio.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Directive, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, ViewChild } from '@angular/core';
import { CustomValueAccessor, customValueAccessorFactory } from '../../../misc/util/internal';
/**
 * @template T
 */
export class FuiRadio {
    constructor() {
        this.isChecked = false;
        this.onCurrentValueChange = new EventEmitter();
        this.onTouched = new EventEmitter();
        this.isDisabled = false;
        this.isReadonly = false;
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get checkedAttribute() {
        return this.isChecked ? '' : undefined;
    }
    /**
     * @return {?}
     */
    get isDisabledAttribute() {
        return this.isDisabled ? 'disabled' : undefined;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onMouseDown(e) {
        e.preventDefault();
    }
    /**
     * @return {?}
     */
    onClick() {
        if (!this.isDisabled && !this.isReadonly) {
            this.currentValue = this.value;
            this.onCurrentValueChange.emit(this.currentValue);
            this.update();
            this.focusRadio();
        }
    }
    /**
     * @return {?}
     */
    onFocusOut() {
        this.onTouched.emit();
    }
    /**
     * @return {?}
     */
    update() {
        this.isChecked = this.currentValue === this.value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.currentValue = value;
        this.update();
    }
    /**
     * @private
     * @return {?}
     */
    focusRadio() {
        this._radioElement.nativeElement.focus();
    }
}
FuiRadio.decorators = [
    { type: Component, args: [{
                selector: 'fui-radio-button',
                template: `
<input class="hidden"
       type="checkbox"
       [attr.name]="name"
       [attr.checked]="checkedAttribute"
       [attr.disabled]="isDisabledAttribute"
       [ngModel]="isChecked"
       (ngModel)="currentValue = value"
       #radio>
<label>
    <ng-content></ng-content>
</label>
`
            }] }
];
/** @nocollapse */
FuiRadio.ctorParameters = () => [];
FuiRadio.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.radio',] }, { type: HostBinding, args: ['class.checkbox',] }],
    name: [{ type: Input }],
    value: [{ type: Input }],
    isChecked: [{ type: HostBinding, args: ['class.checked',] }],
    onCurrentValueChange: [{ type: Output, args: ['currentValueChange',] }],
    onTouched: [{ type: Output, args: ['touched',] }],
    isDisabled: [{ type: Input }],
    isReadonly: [{ type: HostBinding, args: ['class.read-only',] }, { type: Input }],
    _radioElement: [{ type: ViewChild, args: ['radio', { static: false },] }],
    onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
    onClick: [{ type: HostListener, args: ['click',] }],
    onFocusOut: [{ type: HostListener, args: ['focusout',] }]
};
if (false) {
    /** @type {?} */
    FuiRadio.prototype.hasClasses;
    /** @type {?} */
    FuiRadio.prototype.name;
    /** @type {?} */
    FuiRadio.prototype.value;
    /** @type {?} */
    FuiRadio.prototype.isChecked;
    /** @type {?} */
    FuiRadio.prototype.currentValue;
    /** @type {?} */
    FuiRadio.prototype.onCurrentValueChange;
    /** @type {?} */
    FuiRadio.prototype.onTouched;
    /** @type {?} */
    FuiRadio.prototype.isDisabled;
    /** @type {?} */
    FuiRadio.prototype.isReadonly;
    /**
     * @type {?}
     * @private
     */
    FuiRadio.prototype._radioElement;
}
/**
 * @template T
 */
export class FuiRadioValueAccessor extends CustomValueAccessor {
    /**
     * @param {?} host
     */
    constructor(host) {
        super(host);
    }
}
FuiRadioValueAccessor.decorators = [
    { type: Directive, args: [{
                selector: 'fui-radio-button',
                host: {
                    '(currentValueChange)': 'onChange($event)',
                    '(touched)': 'onTouched()'
                },
                providers: [customValueAccessorFactory(FuiRadioValueAccessor)]
            },] }
];
/** @nocollapse */
FuiRadioValueAccessor.ctorParameters = () => [
    { type: FuiRadio }
];
