// ngx-fomantic-ui/esm2015/modules/select/components/multi-select-label.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, Renderer2, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { FuiTransition, Transition, TransitionController, TransitionDirection } from '../../transition/internal';
import { HandledEvent, FuiComponentFactory } from '../../../misc/util/internal';
// See https://github.com/Microsoft/TypeScript/issues/13449.
/** @type {?} */
const templateRef = TemplateRef;
/**
 * @template T
 */
export class FuiMultiSelectLabel extends FuiTransition {
    /**
     * @param {?} renderer
     * @param {?} element
     * @param {?} changeDetector
     * @param {?} componentFactory
     */
    constructor(renderer, element, changeDetector, componentFactory) {
        super(renderer, element, changeDetector);
        this.componentFactory = componentFactory;
        // Initialise transition functionality.
        this._transitionController = new TransitionController(false, 'inline-block');
        this.setTransitionController(this._transitionController);
        this.onDeselected = new EventEmitter();
        this.hasClasses = true;
        this._transitionController.animate(new Transition('scale', 100, TransitionDirection.In));
    }
    /**
     * @return {?}
     */
    get template() {
        return this._template;
    }
    /**
     * @param {?} template
     * @return {?}
     */
    set template(template) {
        this._template = template;
        if (this.template) {
            this.componentFactory.createView(this.templateSibling, this.template, {
                $implicit: this.value,
                query: this.query
            });
        }
    }
    /**
     * @param {?} e
     * @return {?}
     */
    deselectOption(e) {
        e.eventHandled = true;
        this._transitionController.animate(new Transition('scale', 100, TransitionDirection.Out, (/**
         * @return {?}
         */
        () => this.onDeselected.emit(this.value))));
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onClick(e) {
        e.eventHandled = true;
    }
}
FuiMultiSelectLabel.decorators = [
    { type: Component, args: [{
                selector: 'fui-multi-select-label',
                template: `
<span #templateSibling></span>
<span *ngIf="!template" [innerHTML]="formatter(value)"></span>
<i class="delete icon" (click)="deselectOption($event)"></i>
`
            }] }
];
/** @nocollapse */
FuiMultiSelectLabel.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: ChangeDetectorRef },
    { type: FuiComponentFactory }
];
FuiMultiSelectLabel.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.label',] }],
    value: [{ type: Input }],
    query: [{ type: Input }],
    onDeselected: [{ type: Output, args: ['deselected',] }],
    formatter: [{ type: Input }],
    templateSibling: [{ type: ViewChild, args: ['templateSibling', { read: ViewContainerRef, static: false },] }],
    template: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    FuiMultiSelectLabel.prototype.hasClasses;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.value;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.query;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.onDeselected;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.formatter;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.templateSibling;
    /**
     * @type {?}
     * @private
     */
    FuiMultiSelectLabel.prototype._transitionController;
    /**
     * @type {?}
     * @private
     */
    FuiMultiSelectLabel.prototype._template;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.componentFactory;
}
