// ngx-fomantic-ui/esm2015/modules/select/components/select-option.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, Renderer2, ViewChild, ViewContainerRef } from '@angular/core';
import { FuiDropdownMenuItem } from '../../dropdown/internal';
import { HandledEvent } from '../../../misc/util/internal';
/**
 * @template T
 */
export class FuiSelectOption extends FuiDropdownMenuItem {
    /**
     * @param {?} renderer
     * @param {?} element
     * @param {?} changeDetector
     */
    constructor(renderer, element, changeDetector) {
        // We inherit FuiDropdownMenuItem to automatically gain all keyboard navigation functionality.
        // This is not done via adding the .item class because it isn't supported by Angular.
        super(renderer, element);
        this.changeDetector = changeDetector;
        this.hasClasses = true;
        this.isActive = false;
        this.onSelected = new EventEmitter();
        // By default we make the default text an empty label, for the brief moment when it isn't displaying the correct one.
        this.renderedText = '';
        this.usesTemplate = false;
    }
    /**
     * @param {?} formatter
     * @return {?}
     */
    set formatter(formatter) {
        if (!this.usesTemplate) {
            this.renderedText = formatter(this.value);
        }
        else {
            this.renderedText = '';
        }
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onClick(e) {
        e.eventHandled = true;
        setTimeout((/**
         * @return {?}
         */
        () => this.onSelected.emit(this.value)));
    }
}
FuiSelectOption.decorators = [
    { type: Component, args: [{
                selector: 'fui-select-option',
                template: `
<span #templateSibling></span>
<span [innerHTML]="renderedText"></span>
`
            }] }
];
/** @nocollapse */
FuiSelectOption.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
FuiSelectOption.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.item',] }],
    value: [{ type: Input }],
    onSelected: [{ type: Output }],
    isActive: [{ type: HostBinding, args: ['class.active',] }],
    templateSibling: [{ type: ViewChild, args: ['templateSibling', { read: ViewContainerRef, static: false },] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    FuiSelectOption.prototype.hasClasses;
    /** @type {?} */
    FuiSelectOption.prototype.value;
    /** @type {?} */
    FuiSelectOption.prototype.onSelected;
    /** @type {?} */
    FuiSelectOption.prototype.isActive;
    /** @type {?} */
    FuiSelectOption.prototype.renderedText;
    /** @type {?} */
    FuiSelectOption.prototype.usesTemplate;
    /** @type {?} */
    FuiSelectOption.prototype.templateSibling;
    /** @type {?} */
    FuiSelectOption.prototype.changeDetector;
}
