// ngx-fomantic-ui/esm2015/collections/message/components/message-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { ActiveMessage } from '../classes/active-message';
import { FuiMessage } from './message';
import { FuiComponentFactory } from '../../../misc/util/internal';
import { MessageController } from '../classes/message-controller';
export class FuiMessageContainer {
    /**
     * @param {?} _componentFactory
     * @param {?} _element
     */
    constructor(_componentFactory, _element) {
        this._componentFactory = _componentFactory;
        this._element = _element;
        this._messages = [];
        this._queue = [];
    }
    /**
     * @param {?} controller
     * @return {?}
     */
    set controller(controller) {
        controller.registerContainer(this);
    }
    /**
     * @param {?} config
     * @param {?} maxShown
     * @param {?} showNewestFirst
     * @return {?}
     */
    show(config, maxShown, showNewestFirst) {
        /** @type {?} */
        const componentRef = this._componentFactory.createComponent(FuiMessage);
        componentRef.instance.loadConfig(config);
        /** @type {?} */
        const active = new ActiveMessage(config, componentRef)
            .onDismiss((/**
         * @return {?}
         */
        () => this.onMessageClose(active, showNewestFirst)));
        if (this._messages.length < maxShown) {
            this.open(active, showNewestFirst);
        }
        else {
            this.queue(active);
        }
        return active;
    }
    /**
     * @return {?}
     */
    dismissAll() {
        this._queue = [];
        this._messages.forEach((/**
         * @param {?} m
         * @return {?}
         */
        m => m.dismiss()));
    }
    /**
     * @private
     * @param {?} message
     * @param {?} showNewestFirst
     * @return {?}
     */
    open(message, showNewestFirst) {
        this._messages.push(message);
        this._componentFactory.attachToView(message.componentRef, this.containerSibling);
        if (!showNewestFirst) {
            this._componentFactory.moveToElement(message.componentRef, this._element.nativeElement);
        }
        message.component.show();
    }
    /**
     * @private
     * @param {?} message
     * @return {?}
     */
    queue(message) {
        this._queue.push(message);
    }
    /**
     * @private
     * @param {?} message
     * @param {?} showNewestFirst
     * @return {?}
     */
    onMessageClose(message, showNewestFirst) {
        this._messages = this._messages.filter((/**
         * @param {?} m
         * @return {?}
         */
        m => m !== message));
        if (this._queue.length > 0) {
            const [queued] = this._queue.slice(0, 1);
            this.open(queued, showNewestFirst);
        }
    }
}
FuiMessageContainer.decorators = [
    { type: Component, args: [{
                selector: 'fui-message-container',
                template: `
<div #containerSibling></div>
`,
                styles: [`
:host {
    display: block;
}
:host >>> fui-message {
    display: block;
    margin-bottom: 1rem;
}
:host >>> fui-message:last-of-type {
    margin-bottom: 0;
}
:host >>> fui-message {
    cursor: pointer;
}
`]
            }] }
];
/** @nocollapse */
FuiMessageContainer.ctorParameters = () => [
    { type: FuiComponentFactory },
    { type: ElementRef }
];
FuiMessageContainer.propDecorators = {
    containerSibling: [{ type: ViewChild, args: ['containerSibling', { read: ViewContainerRef, static: false },] }],
    controller: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FuiMessageContainer.prototype.containerSibling;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._messages;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._queue;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._componentFactory;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._element;
}
