// ngx-fomantic-ui/esm2015/collections/message/components/message-global-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostListener } from '@angular/core';
import { Util } from '../../../misc/util/internal';
/** @type {?} */
export const MessagePosition = {
    Top: (/** @type {?} */ ('top')),
    TopLeft: (/** @type {?} */ ('top-left')),
    TopRight: (/** @type {?} */ ('top-right')),
    Bottom: (/** @type {?} */ ('bottom')),
    BottomLeft: (/** @type {?} */ ('bottom-left')),
    BottomRight: (/** @type {?} */ ('bottom-right'))
};
export class FuiMessageGlobalContainer {
    /**
     * @return {?}
     */
    get dynamicClasses() {
        /** @type {?} */
        const classes = {};
        this.position
            .split('-')
            .forEach((/**
         * @param {?} p
         * @return {?}
         */
        p => classes[p] = true));
        return classes;
    }
    /**
     * @return {?}
     */
    get dynamicWidth() {
        /** @type {?} */
        const margin = Util.DOM.getDocumentFontSize();
        /** @type {?} */
        let width = this.width;
        if (this.position === MessagePosition.Top ||
            this.position === MessagePosition.Bottom ||
            window.innerWidth < width + margin * 2) {
            width = window.innerWidth - margin * 2;
        }
        return width;
    }
    /**
     * @return {?}
     */
    onDocumentResize() {
    }
}
FuiMessageGlobalContainer.decorators = [
    { type: Component, args: [{
                selector: 'fui-message-global-container',
                template: `
<div class="global container" [ngClass]="dynamicClasses" [style.width.px]="dynamicWidth">
    <fui-message-container [controller]="controller"></fui-message-container>
</div>
`,
                styles: [`
.global.container {
    display: block;
    position: fixed;
}
.global.container.top {
    top: 1rem;
}
.global.container.bottom {
    bottom: 1rem;
}
.global.container.left {
    left: 1rem;
}
.global.container.right {
    right: 1rem;
}
.global.container:not(.left):not(.right) {
    left: 1rem;
}
`]
            }] }
];
FuiMessageGlobalContainer.propDecorators = {
    onDocumentResize: [{ type: HostListener, args: ['window:resize',] }]
};
if (false) {
    /** @type {?} */
    FuiMessageGlobalContainer.prototype.controller;
    /** @type {?} */
    FuiMessageGlobalContainer.prototype.position;
    /** @type {?} */
    FuiMessageGlobalContainer.prototype.width;
}
