// ngx-fomantic-ui/esm5/modules/datepicker/directives/calendar-item.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, EventEmitter, HostBinding, HostListener, Input } from '@angular/core';
var CalendarItem = /** @class */ (function () {
    function CalendarItem(date) {
        this.date = date;
    }
    return CalendarItem;
}());
export { CalendarItem };
if (false) {
    /** @type {?} */
    CalendarItem.prototype.date;
    /** @type {?} */
    CalendarItem.prototype.humanReadable;
    /** @type {?} */
    CalendarItem.prototype.isDisabled;
    /** @type {?} */
    CalendarItem.prototype.isActive;
    /** @type {?} */
    CalendarItem.prototype.isOutsideRange;
    /** @type {?} */
    CalendarItem.prototype.isToday;
    /** @type {?} */
    CalendarItem.prototype.isSelectable;
}
var FuiCalendarItem = /** @class */ (function () {
    function FuiCalendarItem(changeDetector) {
        this.changeDetector = changeDetector;
        this.hasFocus = false;
        this.onFocussed = new EventEmitter();
    }
    Object.defineProperty(FuiCalendarItem.prototype, "isSelectable", {
        get: /**
         * @return {?}
         */
        function () {
            return this.item.isSelectable;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiCalendarItem.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this.item.isActive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiCalendarItem.prototype, "isToday", {
        get: /**
         * @return {?}
         */
        function () {
            return this.item.isToday;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FuiCalendarItem.prototype.onMouseMove = /**
     * @return {?}
     */
    function () {
        if (!this.hasFocus) {
            this.hasFocus = true;
            this.onFocussed.emit(this.hasFocus);
        }
    };
    /**
     * @return {?}
     */
    FuiCalendarItem.prototype.onMouseLeave = /**
     * @return {?}
     */
    function () {
        this.hasFocus = false;
        this.onFocussed.emit(this.hasFocus);
    };
    FuiCalendarItem.decorators = [
        { type: Directive, args: [{
                    selector: '[calendarItem]'
                },] }
    ];
    /** @nocollapse */
    FuiCalendarItem.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    FuiCalendarItem.propDecorators = {
        item: [{ type: Input, args: ['calendarItem',] }],
        hasFocus: [{ type: HostBinding, args: ['class.focus',] }],
        isSelectable: [{ type: HostBinding, args: ['class.disabled',] }],
        isActive: [{ type: HostBinding, args: ['class.active',] }],
        isToday: [{ type: HostBinding, args: ['class.today',] }],
        onMouseMove: [{ type: HostListener, args: ['mousemove',] }],
        onMouseLeave: [{ type: HostListener, args: ['mouseleave',] }]
    };
    return FuiCalendarItem;
}());
export { FuiCalendarItem };
if (false) {
    /** @type {?} */
    FuiCalendarItem.prototype.item;
    /** @type {?} */
    FuiCalendarItem.prototype.hasFocus;
    /** @type {?} */
    FuiCalendarItem.prototype.onFocussed;
    /** @type {?} */
    FuiCalendarItem.prototype.changeDetector;
}
