// ngx-fomantic-ui/esm2015/collections/message/services/message-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FuiComponentFactory } from '../../../misc/util/internal';
import { MessagePosition, FuiMessageGlobalContainer } from '../components/message-global-container';
import { MessageController } from '../classes/message-controller';
export class FuiMessageService {
    /**
     * @param {?} _componentFactory
     */
    constructor(_componentFactory) {
        this._componentFactory = _componentFactory;
        this._controller = new MessageController();
        this._containerRef = this._componentFactory.createComponent(FuiMessageGlobalContainer);
        this._container.controller = this._controller;
        this._componentFactory.attachToApplication(this._containerRef);
        this._componentFactory.moveToDocumentBody(this._containerRef);
        this.position = MessagePosition.TopRight;
        this.width = 480;
    }
    /**
     * @return {?}
     */
    get position() {
        return this._container.position;
    }
    /**
     * @param {?} position
     * @return {?}
     */
    set position(position) {
        this._container.position = position;
    }
    /**
     * @return {?}
     */
    get width() {
        return this._container.width;
    }
    /**
     * @param {?} width
     * @return {?}
     */
    set width(width) {
        this._container.width = width;
    }
    /**
     * @return {?}
     */
    get maxShown() {
        return this._controller.maxShown;
    }
    /**
     * @param {?} max
     * @return {?}
     */
    set maxShown(max) {
        this._controller.maxShown = max;
    }
    /**
     * @return {?}
     */
    get isNewestOnTop() {
        return this._controller.isNewestOnTop;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isNewestOnTop(value) {
        this._controller.isNewestOnTop = value;
    }
    /**
     * @private
     * @return {?}
     */
    get _container() {
        return this._containerRef.instance;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    show(config) {
        return this._controller.show(config);
    }
    /**
     * @return {?}
     */
    dismissAll() {
        return this._controller.dismissAll();
    }
}
FuiMessageService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FuiMessageService.ctorParameters = () => [
    { type: FuiComponentFactory }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FuiMessageService.prototype._controller;
    /**
     * @type {?}
     * @private
     */
    FuiMessageService.prototype._containerRef;
    /**
     * @type {?}
     * @private
     */
    FuiMessageService.prototype._componentFactory;
}
