// ngx-fomantic-ui/esm5/collections/message/classes/active-message.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
FuiActiveMessage = /** @class */ (function () {
    function FuiActiveMessage() {
    }
    return FuiActiveMessage;
}());
/**
 * @abstract
 */
export { FuiActiveMessage };
if (false) {
    /**
     * @abstract
     * @param {?} callback
     * @return {?}
     */
    FuiActiveMessage.prototype.onClick = function (callback) { };
    /**
     * @abstract
     * @param {?} callback
     * @return {?}
     */
    FuiActiveMessage.prototype.onDismiss = function (callback) { };
    /**
     * @abstract
     * @return {?}
     */
    FuiActiveMessage.prototype.dismiss = function () { };
}
var ActiveMessage = /** @class */ (function () {
    function ActiveMessage(config, componentRef) {
        var _this = this;
        this.config = config;
        this.componentRef = componentRef;
        this.component.onDismiss.subscribe((/**
         * @return {?}
         */
        function () { return _this.componentRef.destroy(); }));
    }
    Object.defineProperty(ActiveMessage.prototype, "component", {
        get: /**
         * @return {?}
         */
        function () {
            return this.componentRef.instance;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} callback
     * @return {?}
     */
    ActiveMessage.prototype.onClick = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        this.config.onClick.subscribe((/**
         * @return {?}
         */
        function () { return callback(); }));
        return this;
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    ActiveMessage.prototype.onDismiss = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        this.config.onDismiss.subscribe((/**
         * @return {?}
         */
        function () { return callback(); }));
        return this;
    };
    /**
     * @return {?}
     */
    ActiveMessage.prototype.dismiss = /**
     * @return {?}
     */
    function () {
        this.component.dismiss();
    };
    return ActiveMessage;
}());
export { ActiveMessage };
if (false) {
    /** @type {?} */
    ActiveMessage.prototype.config;
    /** @type {?} */
    ActiveMessage.prototype.componentRef;
}
