// ngx-fomantic-ui/esm5/collections/message/classes/message-controller.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IMessageController() { }
if (false) {
    /** @type {?} */
    IMessageController.prototype.maxShown;
    /** @type {?} */
    IMessageController.prototype.isNewestOnTop;
    /**
     * @param {?} config
     * @return {?}
     */
    IMessageController.prototype.show = function (config) { };
    /**
     * @return {?}
     */
    IMessageController.prototype.dismissAll = function () { };
}
var MessageController = /** @class */ (function () {
    function MessageController() {
        this.maxShown = 7;
        this.isNewestOnTop = true;
    }
    /**
     * @param {?} container
     * @return {?}
     */
    MessageController.prototype.registerContainer = /**
     * @param {?} container
     * @return {?}
     */
    function (container) {
        this._container = container;
    };
    /**
     * @param {?} config
     * @return {?}
     */
    MessageController.prototype.show = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        this.throwContainerError();
        return this._container.show(config, this.maxShown, this.isNewestOnTop);
    };
    /**
     * @return {?}
     */
    MessageController.prototype.dismissAll = /**
     * @return {?}
     */
    function () {
        this.throwContainerError();
        return this._container.dismissAll();
    };
    /**
     * @private
     * @return {?}
     */
    MessageController.prototype.throwContainerError = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this._container) {
            throw new Error('You must pass this controller to a message container.');
        }
    };
    return MessageController;
}());
export { MessageController };
if (false) {
    /** @type {?} */
    MessageController.prototype.maxShown;
    /** @type {?} */
    MessageController.prototype.isNewestOnTop;
    /**
     * @type {?}
     * @private
     */
    MessageController.prototype._container;
}
