// ngx-fomantic-ui/esm5/collections/message/classes/message-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
/** @type {?} */
export var MessageState = {
    Default: (/** @type {?} */ ('')),
    Info: (/** @type {?} */ ('info')),
    Success: (/** @type {?} */ ('success')),
    Warning: (/** @type {?} */ ('warning')),
    Error: (/** @type {?} */ ('error'))
};
var MessageConfig = /** @class */ (function () {
    function MessageConfig(text, state, header) {
        if (state === void 0) { state = MessageState.Default; }
        this.text = text;
        this.state = state;
        this.header = header;
        this.timeout = 5000;
        this.extendedTimeout = 1000;
        this.hasDismissButton = true;
        this.hasProgress = false;
        this.transition = 'fade';
        this.transitionInDuration = 400;
        this.transitionOutDuration = 1000;
        this.onClick = new EventEmitter();
        this.onDismiss = new EventEmitter();
    }
    return MessageConfig;
}());
export { MessageConfig };
if (false) {
    /** @type {?} */
    MessageConfig.prototype.text;
    /** @type {?} */
    MessageConfig.prototype.header;
    /** @type {?} */
    MessageConfig.prototype.state;
    /** @type {?} */
    MessageConfig.prototype.timeout;
    /** @type {?} */
    MessageConfig.prototype.extendedTimeout;
    /** @type {?} */
    MessageConfig.prototype.hasDismissButton;
    /** @type {?} */
    MessageConfig.prototype.hasProgress;
    /** @type {?} */
    MessageConfig.prototype.transition;
    /** @type {?} */
    MessageConfig.prototype.transitionInDuration;
    /** @type {?} */
    MessageConfig.prototype.transitionOutDuration;
    /** @type {?} */
    MessageConfig.prototype.onClick;
    /** @type {?} */
    MessageConfig.prototype.onDismiss;
}
