// ngx-fomantic-ui/esm5/modules/datepicker/components/calendar-view-title.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { CalendarRangeService } from '../services/calendar-range.service';
var FuiCalendarViewTitle = /** @class */ (function () {
    function FuiCalendarViewTitle() {
        this.onZoomOut = new EventEmitter();
    }
    FuiCalendarViewTitle.decorators = [
        { type: Component, args: [{
                    selector: 'fui-calendar-view-title',
                    template: "\n<span class=\"title link\" (click)=\"onZoomOut.emit()\">\n    <ng-content></ng-content>\n</span>\n<span class=\"prev link\" [class.disabled]=\"!ranges?.canMovePrevious\" (click)=\"ranges?.movePrevious()\">\n    <i class=\"chevron left icon\"></i>\n</span>\n<span class=\"next link\" [class.disabled]=\"!ranges?.canMoveNext\" (click)=\"ranges?.moveNext()\">\n    <i class=\"chevron right icon\"></i>\n</span>\n",
                    styles: ["\n.title.link {\n    display: inline-block;\n    margin-left: 2rem;\n    margin-right: 2rem;\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiCalendarViewTitle.ctorParameters = function () { return []; };
    FuiCalendarViewTitle.propDecorators = {
        ranges: [{ type: Input }],
        onZoomOut: [{ type: Output, args: ['zoomOut',] }]
    };
    return FuiCalendarViewTitle;
}());
export { FuiCalendarViewTitle };
if (false) {
    /** @type {?} */
    FuiCalendarViewTitle.prototype.ranges;
    /** @type {?} */
    FuiCalendarViewTitle.prototype.onZoomOut;
}
