// ngx-fomantic-ui/esm5/modules/datepicker/components/datepicker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, HostListener } from '@angular/core';
import { CalendarService } from './../services/calendar.service';
import { DatetimeConfig } from '../classes/calendar-config';
import { FuiLocalizationService } from '../../../behaviors/localization/internal';
/** @type {?} */
export var DatepickerMode = {
    Year: (/** @type {?} */ ('year')),
    Month: (/** @type {?} */ ('month')),
    Date: (/** @type {?} */ ('date')),
    Datetime: (/** @type {?} */ ('datetime')),
    Time: (/** @type {?} */ ('time'))
};
var FuiDatepicker = /** @class */ (function () {
    function FuiDatepicker(localizationService) {
        this.service = new CalendarService(new DatetimeConfig(), localizationService.get().datepicker);
        this.hasClasses = true;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    FuiDatepicker.prototype.onMouseDown = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.preventDefault();
    };
    FuiDatepicker.decorators = [
        { type: Component, args: [{
                    selector: 'fui-datepicker',
                    template: "\n<ng-container [ngSwitch]=\"service.currentView\">\n    <fui-calendar-year-view [service]=\"service\" *ngSwitchCase=\"0\"></fui-calendar-year-view>\n    <fui-calendar-month-view [service]=\"service\" *ngSwitchCase=\"1\"></fui-calendar-month-view>\n    <fui-calendar-date-view [service]=\"service\" *ngSwitchCase=\"2\"></fui-calendar-date-view>\n    <fui-calendar-hour-view [service]=\"service\" *ngSwitchCase=\"3\"></fui-calendar-hour-view>\n    <fui-calendar-minute-view [service]=\"service\" *ngSwitchCase=\"4\"></fui-calendar-minute-view>\n</ng-container>\n",
                    styles: ["\n:host {\n    user-select: none;\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiDatepicker.ctorParameters = function () { return [
        { type: FuiLocalizationService }
    ]; };
    FuiDatepicker.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.active',] }, { type: HostBinding, args: ['class.calendar',] }],
        onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
    };
    return FuiDatepicker;
}());
export { FuiDatepicker };
if (false) {
    /** @type {?} */
    FuiDatepicker.prototype.hasClasses;
    /** @type {?} */
    FuiDatepicker.prototype.service;
}
