// ngx-fomantic-ui/esm5/modules/search/components/search-result.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Input, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { FuiComponentFactory } from '../../../misc/util/internal';
// See https://github.com/Microsoft/TypeScript/issues/13449.
/** @type {?} */
var templateRef = TemplateRef;
/**
 * @template T
 */
var FuiSearchResult = /** @class */ (function () {
    function FuiSearchResult(componentFactory) {
        this.componentFactory = componentFactory;
        this.hasClasses = true;
        // By default we make this function return an empty string, for the brief moment when it isn't displaying the correct label.
        this.formatter = (/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return ''; });
    }
    Object.defineProperty(FuiSearchResult.prototype, "template", {
        get: /**
         * @return {?}
         */
        function () {
            return this._template;
        },
        set: /**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            this._template = template;
            if (this.template) {
                this.componentFactory.createView(this.templateSibling, this.template, {
                    $implicit: this.value,
                    query: this.query
                });
            }
        },
        enumerable: true,
        configurable: true
    });
    FuiSearchResult.decorators = [
        { type: Component, args: [{
                    selector: 'fui-search-result',
                    template: "\n<span #templateSibling></span>\n<span *ngIf=\"!template\" [innerHTML]=\"formatter(value, query)\"></span>\n"
                }] }
    ];
    /** @nocollapse */
    FuiSearchResult.ctorParameters = function () { return [
        { type: FuiComponentFactory }
    ]; };
    FuiSearchResult.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.result',] }],
        value: [{ type: Input }],
        query: [{ type: Input }],
        formatter: [{ type: Input }],
        templateSibling: [{ type: ViewChild, args: ['templateSibling', { read: ViewContainerRef, static: false },] }],
        template: [{ type: Input }]
    };
    return FuiSearchResult;
}());
export { FuiSearchResult };
if (false) {
    /** @type {?} */
    FuiSearchResult.prototype.hasClasses;
    /** @type {?} */
    FuiSearchResult.prototype.value;
    /** @type {?} */
    FuiSearchResult.prototype.query;
    /** @type {?} */
    FuiSearchResult.prototype.formatter;
    /** @type {?} */
    FuiSearchResult.prototype.templateSibling;
    /**
     * @type {?}
     * @private
     */
    FuiSearchResult.prototype._template;
    /** @type {?} */
    FuiSearchResult.prototype.componentFactory;
}
