// ngx-fomantic-ui/esm5/modules/checkbox/components/checkbox.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Directive, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, ViewChild } from '@angular/core';
import { CustomValueAccessor, customValueAccessorFactory } from '../../../misc/util/internal';
var FuiCheckbox = /** @class */ (function () {
    function FuiCheckbox() {
        this.isChecked = false;
        this.onCheckChange = new EventEmitter();
        this.onTouched = new EventEmitter();
        this.isDisabled = false;
        this.isReadonly = false;
        this.hasClasses = true;
    }
    Object.defineProperty(FuiCheckbox.prototype, "checkedAttribute", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isChecked ? '' : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiCheckbox.prototype, "isDisabledAttribute", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isDisabled ? 'disabled' : undefined;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} e
     * @return {?}
     */
    FuiCheckbox.prototype.onMouseDown = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.preventDefault();
    };
    /**
     * @return {?}
     */
    FuiCheckbox.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (!this.isDisabled && !this.isReadonly) {
            this.toggle();
            this.focusCheckbox();
        }
    };
    /**
     * @return {?}
     */
    FuiCheckbox.prototype.onFocusOut = /**
     * @return {?}
     */
    function () {
        this.onTouched.emit();
    };
    /**
     * @return {?}
     */
    FuiCheckbox.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this.isChecked = !this.isChecked;
        this.onCheckChange.emit(this.isChecked);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FuiCheckbox.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.isChecked = value;
    };
    /**
     * @private
     * @return {?}
     */
    FuiCheckbox.prototype.focusCheckbox = /**
     * @private
     * @return {?}
     */
    function () {
        this._checkboxElement.nativeElement.focus();
    };
    FuiCheckbox.decorators = [
        { type: Component, args: [{
                    selector: 'fui-checkbox',
                    exportAs: 'fuiCheckbox',
                    template: "\n<input class=\"hidden\"\n       type=\"checkbox\"\n       [attr.name]=\"name\"\n       [attr.checked]=\"checkedAttribute\"\n       [attr.disabled]=\"isDisabledAttribute\"\n       [(ngModel)]=\"isChecked\"\n       #checkbox>\n<label>\n    <ng-content></ng-content>\n</label>\n"
                }] }
    ];
    /** @nocollapse */
    FuiCheckbox.ctorParameters = function () { return []; };
    FuiCheckbox.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.checkbox',] }],
        name: [{ type: Input }],
        isChecked: [{ type: HostBinding, args: ['class.checked',] }],
        onCheckChange: [{ type: Output, args: ['checkChange',] }],
        onTouched: [{ type: Output, args: ['touched',] }],
        isDisabled: [{ type: Input }],
        isReadonly: [{ type: HostBinding, args: ['class.read-only',] }, { type: Input }],
        _checkboxElement: [{ type: ViewChild, args: ['checkbox', { static: false },] }],
        onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
        onClick: [{ type: HostListener, args: ['click',] }],
        onFocusOut: [{ type: HostListener, args: ['focusout',] }]
    };
    return FuiCheckbox;
}());
export { FuiCheckbox };
if (false) {
    /** @type {?} */
    FuiCheckbox.prototype.hasClasses;
    /** @type {?} */
    FuiCheckbox.prototype.name;
    /** @type {?} */
    FuiCheckbox.prototype.isChecked;
    /** @type {?} */
    FuiCheckbox.prototype.onCheckChange;
    /** @type {?} */
    FuiCheckbox.prototype.onTouched;
    /** @type {?} */
    FuiCheckbox.prototype.isDisabled;
    /** @type {?} */
    FuiCheckbox.prototype.isReadonly;
    /**
     * @type {?}
     * @private
     */
    FuiCheckbox.prototype._checkboxElement;
}
var FuiCheckboxValueAccessor = /** @class */ (function (_super) {
    tslib_1.__extends(FuiCheckboxValueAccessor, _super);
    function FuiCheckboxValueAccessor(host) {
        return _super.call(this, host) || this;
    }
    FuiCheckboxValueAccessor.decorators = [
        { type: Directive, args: [{
                    selector: 'fui-checkbox',
                    host: {
                        '(checkChange)': 'onChange($event)',
                        '(touched)': 'onTouched()'
                    },
                    providers: [customValueAccessorFactory(FuiCheckboxValueAccessor)]
                },] }
    ];
    /** @nocollapse */
    FuiCheckboxValueAccessor.ctorParameters = function () { return [
        { type: FuiCheckbox }
    ]; };
    return FuiCheckboxValueAccessor;
}(CustomValueAccessor));
export { FuiCheckboxValueAccessor };
