// ngx-fomantic-ui/esm5/modules/checkbox/components/radio.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Directive, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, ViewChild } from '@angular/core';
import { CustomValueAccessor, customValueAccessorFactory } from '../../../misc/util/internal';
/**
 * @template T
 */
var FuiRadio = /** @class */ (function () {
    function FuiRadio() {
        this.isChecked = false;
        this.onCurrentValueChange = new EventEmitter();
        this.onTouched = new EventEmitter();
        this.isDisabled = false;
        this.isReadonly = false;
        this.hasClasses = true;
    }
    Object.defineProperty(FuiRadio.prototype, "checkedAttribute", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isChecked ? '' : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiRadio.prototype, "isDisabledAttribute", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isDisabled ? 'disabled' : undefined;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} e
     * @return {?}
     */
    FuiRadio.prototype.onMouseDown = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.preventDefault();
    };
    /**
     * @return {?}
     */
    FuiRadio.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (!this.isDisabled && !this.isReadonly) {
            this.currentValue = this.value;
            this.onCurrentValueChange.emit(this.currentValue);
            this.update();
            this.focusRadio();
        }
    };
    /**
     * @return {?}
     */
    FuiRadio.prototype.onFocusOut = /**
     * @return {?}
     */
    function () {
        this.onTouched.emit();
    };
    /**
     * @return {?}
     */
    FuiRadio.prototype.update = /**
     * @return {?}
     */
    function () {
        this.isChecked = this.currentValue === this.value;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FuiRadio.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.currentValue = value;
        this.update();
    };
    /**
     * @private
     * @return {?}
     */
    FuiRadio.prototype.focusRadio = /**
     * @private
     * @return {?}
     */
    function () {
        this._radioElement.nativeElement.focus();
    };
    FuiRadio.decorators = [
        { type: Component, args: [{
                    selector: 'fui-radio-button',
                    template: "\n<input class=\"hidden\"\n       type=\"checkbox\"\n       [attr.name]=\"name\"\n       [attr.checked]=\"checkedAttribute\"\n       [attr.disabled]=\"isDisabledAttribute\"\n       [ngModel]=\"isChecked\"\n       (ngModel)=\"currentValue = value\"\n       #radio>\n<label>\n    <ng-content></ng-content>\n</label>\n"
                }] }
    ];
    /** @nocollapse */
    FuiRadio.ctorParameters = function () { return []; };
    FuiRadio.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.radio',] }, { type: HostBinding, args: ['class.checkbox',] }],
        name: [{ type: Input }],
        value: [{ type: Input }],
        isChecked: [{ type: HostBinding, args: ['class.checked',] }],
        onCurrentValueChange: [{ type: Output, args: ['currentValueChange',] }],
        onTouched: [{ type: Output, args: ['touched',] }],
        isDisabled: [{ type: Input }],
        isReadonly: [{ type: HostBinding, args: ['class.read-only',] }, { type: Input }],
        _radioElement: [{ type: ViewChild, args: ['radio', { static: false },] }],
        onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
        onClick: [{ type: HostListener, args: ['click',] }],
        onFocusOut: [{ type: HostListener, args: ['focusout',] }]
    };
    return FuiRadio;
}());
export { FuiRadio };
if (false) {
    /** @type {?} */
    FuiRadio.prototype.hasClasses;
    /** @type {?} */
    FuiRadio.prototype.name;
    /** @type {?} */
    FuiRadio.prototype.value;
    /** @type {?} */
    FuiRadio.prototype.isChecked;
    /** @type {?} */
    FuiRadio.prototype.currentValue;
    /** @type {?} */
    FuiRadio.prototype.onCurrentValueChange;
    /** @type {?} */
    FuiRadio.prototype.onTouched;
    /** @type {?} */
    FuiRadio.prototype.isDisabled;
    /** @type {?} */
    FuiRadio.prototype.isReadonly;
    /**
     * @type {?}
     * @private
     */
    FuiRadio.prototype._radioElement;
}
/**
 * @template T
 */
var FuiRadioValueAccessor = /** @class */ (function (_super) {
    tslib_1.__extends(FuiRadioValueAccessor, _super);
    function FuiRadioValueAccessor(host) {
        return _super.call(this, host) || this;
    }
    FuiRadioValueAccessor.decorators = [
        { type: Directive, args: [{
                    selector: 'fui-radio-button',
                    host: {
                        '(currentValueChange)': 'onChange($event)',
                        '(touched)': 'onTouched()'
                    },
                    providers: [customValueAccessorFactory(FuiRadioValueAccessor)]
                },] }
    ];
    /** @nocollapse */
    FuiRadioValueAccessor.ctorParameters = function () { return [
        { type: FuiRadio }
    ]; };
    return FuiRadioValueAccessor;
}(CustomValueAccessor));
export { FuiRadioValueAccessor };
