// ngx-fomantic-ui/esm2015/modules/select/directives/select-search.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, HostBinding, HostListener, Renderer2 } from '@angular/core';
export class FuiSelectSearch {
    /**
     * @param {?} _renderer
     * @param {?} _element
     */
    constructor(_renderer, _element) {
        this._renderer = _renderer;
        this._element = _element;
        this.onQueryUpdated = new EventEmitter();
        this.onQueryKeyDown = new EventEmitter();
        this.hasClasses = true;
        this.autoComplete = 'off';
    }
    /**
     * @param {?} query
     * @return {?}
     */
    set query(query) {
        this._renderer.setProperty(this._element.nativeElement, 'value', query);
    }
    /**
     * @param {?} query
     * @return {?}
     */
    updateQuery(query) {
        this.onQueryUpdated.emit(query);
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onKeyDown(e) {
        this.onQueryKeyDown.emit(e);
    }
    /**
     * @return {?}
     */
    focus() {
        // Slightly delay to support in menu search.
        this._element.nativeElement.focus();
        setTimeout((/**
         * @return {?}
         */
        () => this._element.nativeElement.focus()));
    }
}
FuiSelectSearch.decorators = [
    { type: Directive, args: [{
                selector: 'input[fuiSelectSearch]'
            },] }
];
/** @nocollapse */
FuiSelectSearch.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
FuiSelectSearch.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.search',] }],
    autoComplete: [{ type: HostBinding, args: ['attr.autocomplete',] }],
    updateQuery: [{ type: HostListener, args: ['input', ['$event.target.value'],] }],
    onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    FuiSelectSearch.prototype.hasClasses;
    /** @type {?} */
    FuiSelectSearch.prototype.autoComplete;
    /** @type {?} */
    FuiSelectSearch.prototype.onQueryUpdated;
    /** @type {?} */
    FuiSelectSearch.prototype.onQueryKeyDown;
    /**
     * @type {?}
     * @private
     */
    FuiSelectSearch.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FuiSelectSearch.prototype._element;
}
