// ngx-fomantic-ui/esm5/modules/sidebar/components/sidebar-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, HostBinding } from '@angular/core';
import { FuiSidebar } from './sidebar';
import { FuiSidebarSibling } from './sidebar-sibling';
var FuiSidebarContainer = /** @class */ (function () {
    function FuiSidebarContainer() {
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    FuiSidebarContainer.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        if (!this.sidebar) {
            throw new Error('You must include a <fui-sidebar> element within the container.');
        }
        this.service = this.sidebar.service;
        if (!this.sibling) {
            throw new Error('You must include a <fui-sidebar-sibling> element within the container.');
        }
        this.sibling.service = this.service;
    };
    FuiSidebarContainer.decorators = [
        { type: Component, args: [{
                    selector: 'fui-sidebar-container',
                    template: "<ng-content></ng-content>",
                    styles: ["\n:host {\n    display: block;\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiSidebarContainer.ctorParameters = function () { return []; };
    FuiSidebarContainer.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.pushable',] }],
        sidebar: [{ type: ContentChild, args: [FuiSidebar, { static: false },] }],
        sibling: [{ type: ContentChild, args: [FuiSidebarSibling, { static: false },] }]
    };
    return FuiSidebarContainer;
}());
export { FuiSidebarContainer };
if (false) {
    /** @type {?} */
    FuiSidebarContainer.prototype.service;
    /** @type {?} */
    FuiSidebarContainer.prototype.hasClasses;
    /** @type {?} */
    FuiSidebarContainer.prototype.sidebar;
    /** @type {?} */
    FuiSidebarContainer.prototype.sibling;
}
