// ngx-fomantic-ui/esm5/modules/sidebar/components/sidebar.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostBinding, Input, Output, Renderer2 } from '@angular/core';
import { SidebarDirection, SidebarService, SidebarTransition } from '../services/sidebar.service';
var FuiSidebar = /** @class */ (function () {
    function FuiSidebar(_renderer, _element) {
        var _this = this;
        this._renderer = _renderer;
        this._element = _element;
        this.service = new SidebarService();
        // We set the default here as well to force the classes to update.
        this.transition = SidebarTransition.Uncover;
        this.direction = SidebarDirection.Left;
        setTimeout((/**
         * @return {?}
         */
        function () { return _this.updateDimensions(); }));
        this.service.isVisibleChange.subscribe((/**
         * @return {?}
         */
        function () { return _this.updateDimensions(); }));
        this.hasClasses = true;
    }
    Object.defineProperty(FuiSidebar.prototype, "transition", {
        get: /**
         * @return {?}
         */
        function () {
            return this.service.transition;
        },
        set: /**
         * @param {?} transition
         * @return {?}
         */
        function (transition) {
            var _this = this;
            this.service.transition.split(' ').forEach((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return _this.setClass(c, false); }));
            this.service.transition = transition;
            this.service.transition.split(' ').forEach((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return _this.setClass(c, true); }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiSidebar.prototype, "direction", {
        get: /**
         * @return {?}
         */
        function () {
            return this.service.direction;
        },
        set: /**
         * @param {?} direction
         * @return {?}
         */
        function (direction) {
            this.setClass(this.service.direction, false);
            this.service.direction = direction;
            this.setClass(this.service.direction, true);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiSidebar.prototype, "isVisible", {
        get: /**
         * @return {?}
         */
        function () {
            return this.service.isVisible;
        },
        set: /**
         * @param {?} isVisible
         * @return {?}
         */
        function (isVisible) {
            this.service.setVisibleState(isVisible);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiSidebar.prototype, "isVisibleChange", {
        get: /**
         * @return {?}
         */
        function () {
            return this.service.isVisibleChange;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiSidebar.prototype, "isAnimating", {
        get: /**
         * @return {?}
         */
        function () {
            return this.service.isAnimating;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FuiSidebar.prototype.open = /**
     * @return {?}
     */
    function () {
        this.service.setVisibleState(true);
    };
    /**
     * @return {?}
     */
    FuiSidebar.prototype.close = /**
     * @return {?}
     */
    function () {
        this.service.setVisibleState(false);
    };
    /**
     * @return {?}
     */
    FuiSidebar.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this.service.toggleVisibleState();
    };
    /**
     * @private
     * @return {?}
     */
    FuiSidebar.prototype.updateDimensions = /**
     * @private
     * @return {?}
     */
    function () {
        this.service.width = this._element.nativeElement.offsetWidth;
        this.service.height = this._element.nativeElement.offsetHeight;
    };
    /**
     * @private
     * @param {?} className
     * @param {?=} isAdd
     * @return {?}
     */
    FuiSidebar.prototype.setClass = /**
     * @private
     * @param {?} className
     * @param {?=} isAdd
     * @return {?}
     */
    function (className, isAdd) {
        if (isAdd === void 0) { isAdd = true; }
        if (isAdd) {
            this._renderer.addClass(this._element.nativeElement, className);
        }
        else {
            this._renderer.removeClass(this._element.nativeElement, className);
        }
    };
    FuiSidebar.decorators = [
        { type: Component, args: [{
                    selector: 'fui-sidebar',
                    template: "<ng-content></ng-content>"
                }] }
    ];
    /** @nocollapse */
    FuiSidebar.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    FuiSidebar.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.sidebar',] }, { type: HostBinding, args: ['class.menu',] }],
        transition: [{ type: Input }],
        direction: [{ type: Input }],
        isVisible: [{ type: HostBinding, args: ['class.visible',] }, { type: Input }],
        isVisibleChange: [{ type: Output }],
        isAnimating: [{ type: HostBinding, args: ['class.animating',] }]
    };
    return FuiSidebar;
}());
export { FuiSidebar };
if (false) {
    /** @type {?} */
    FuiSidebar.prototype.service;
    /** @type {?} */
    FuiSidebar.prototype.hasClasses;
    /**
     * @type {?}
     * @private
     */
    FuiSidebar.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FuiSidebar.prototype._element;
}
