// ngx-fomantic-ui/esm5/modules/sidebar/components/sidebar-sibling.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostBinding, HostListener, Input, Renderer2 } from '@angular/core';
import { SidebarTransition } from '../services/sidebar.service';
var FuiSidebarSibling = /** @class */ (function () {
    function FuiSidebarSibling(_renderer, _element) {
        this._renderer = _renderer;
        this._element = _element;
        this.canCloseSidebar = true;
        this.isDimmedWhenVisible = false;
        this.hasClasses = true;
    }
    Object.defineProperty(FuiSidebarSibling.prototype, "isVisible", {
        get: /**
         * @return {?}
         */
        function () {
            if (!this.service) {
                return false;
            }
            return this.service.isVisible;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiSidebarSibling.prototype, "isDimmed", {
        get: /**
         * @return {?}
         */
        function () {
            if (!this.service) {
                return false;
            }
            return this.service.isVisible && this.isDimmedWhenVisible;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiSidebarSibling.prototype, "service", {
        get: /**
         * @return {?}
         */
        function () {
            return this._service;
        },
        set: /**
         * @param {?} service
         * @return {?}
         */
        function (service) {
            var _this = this;
            this._service = service;
            setTimeout((/**
             * @return {?}
             */
            function () { return _this.updateTransform(); }));
            this._service.isVisibleChange.subscribe((/**
             * @return {?}
             */
            function () { return _this.updateTransform(); }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    FuiSidebarSibling.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.canCloseSidebar && this.service.isVisible && !this.service.wasJustOpened) {
            this.service.setVisibleState(false);
        }
    };
    /**
     * @private
     * @return {?}
     */
    FuiSidebarSibling.prototype.updateTransform = /**
     * @private
     * @return {?}
     */
    function () {
        this._renderer.removeStyle(this._element.nativeElement, 'transform');
        this._renderer.removeStyle(this._element.nativeElement, '-webkit-transform');
        if (this.service.isVisible &&
            this.service.transition !== SidebarTransition.Overlay &&
            this.service.transition !== SidebarTransition.ScaleDown) {
            /** @type {?} */
            var translate = "translate3d(" + this.service.width + "px, " + this.service.height + "px, 0)";
            this._renderer.setStyle(this._element.nativeElement, 'transform', translate);
            this._renderer.setStyle(this._element.nativeElement, '-webkit-transform', translate);
        }
    };
    FuiSidebarSibling.decorators = [
        { type: Component, args: [{
                    selector: 'fui-sidebar-sibling',
                    template: "\n                   <ng-content></ng-content>",
                    styles: ["\n                   :host {\n                       display: block;\n                   }\n               "]
                }] }
    ];
    /** @nocollapse */
    FuiSidebarSibling.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    FuiSidebarSibling.propDecorators = {
        isDimmedWhenVisible: [{ type: Input }],
        hasClasses: [{ type: HostBinding, args: ['class.pusher',] }],
        canCloseSidebar: [{ type: Input }],
        isVisible: [{ type: HostBinding, args: ['class.visible',] }],
        isDimmed: [{ type: HostBinding, args: ['class.dimmed',] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return FuiSidebarSibling;
}());
export { FuiSidebarSibling };
if (false) {
    /** @type {?} */
    FuiSidebarSibling.prototype.isDimmedWhenVisible;
    /** @type {?} */
    FuiSidebarSibling.prototype.hasClasses;
    /** @type {?} */
    FuiSidebarSibling.prototype.canCloseSidebar;
    /**
     * @type {?}
     * @private
     */
    FuiSidebarSibling.prototype._service;
    /**
     * @type {?}
     * @private
     */
    FuiSidebarSibling.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FuiSidebarSibling.prototype._element;
}
