// ngx-fomantic-ui/esm5/modules/tabs/directives/tab-header.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, HostBinding, HostListener, Input, Output } from '@angular/core';
var FuiTabHeader = /** @class */ (function () {
    function FuiTabHeader() {
        this._isActive = false;
        this.isActiveChange = new EventEmitter();
        this.isActiveExternalChange = new EventEmitter();
        this.onActivate = new EventEmitter();
        this.onDeactivate = new EventEmitter();
        this.isDisabled = false;
        this.hasClasses = true;
    }
    Object.defineProperty(FuiTabHeader.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isActive;
        },
        set: /**
         * @param {?} active
         * @return {?}
         */
        function (active) {
            var _this = this;
            /** @type {?} */
            var isActive = active;
            // Only used by @Input(), runs whenever user input changes `isActive`.
            // Run in timeout because `isDisabled` can prohibit user from changing `isActive`.
            // so update is delayed to avoid 'changed after checked' error.
            setTimeout((/**
             * @return {?}
             */
            function () {
                // Only allow change if tab header is not disabled.
                isActive = !_this.isDisabled ? active : false;
                _this.setActiveState(isActive);
                // Fire 'external change' event as user input has occured.
                _this.isActiveExternalChange.emit(isActive);
            }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiTabHeader.prototype, "isDisabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isDisabled;
        },
        set: /**
         * @param {?} disabled
         * @return {?}
         */
        function (disabled) {
            // Only update if value provided is different to current one.
            if (this._isDisabled !== disabled) {
                this._isDisabled = disabled;
                // If now disabled, then tab header must be deactivated.
                if (this.isDisabled) {
                    this.isActive = false;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    // Internally update active state.
    // Internally update active state.
    /**
     * @param {?} active
     * @return {?}
     */
    FuiTabHeader.prototype.setActiveState = 
    // Internally update active state.
    /**
     * @param {?} active
     * @return {?}
     */
    function (active) {
        // If (cast) active value has changed:
        if (!!this._isActive !== active) {
            // Update to the new value.
            this._isActive = active;
            // Fire the appropriate activation event.
            if (active) {
                this.onActivate.emit();
            }
            else {
                this.onDeactivate.emit();
            }
        }
        // Regardless, emit a change to `isActive`, so [(isActive)] works correctly.
        this.isActiveChange.emit(active);
    };
    /**
     * @return {?}
     */
    FuiTabHeader.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (!this.isDisabled) {
            // Activate the tab when clicked, so long as it isn't disabled.
            this.isActive = true;
        }
    };
    FuiTabHeader.decorators = [
        { type: Directive, args: [{
                    selector: '[fuiTabHeader]'
                },] }
    ];
    /** @nocollapse */
    FuiTabHeader.ctorParameters = function () { return []; };
    FuiTabHeader.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.item',] }],
        id: [{ type: Input, args: ['fuiTabHeader',] }],
        isActiveChange: [{ type: Output }],
        onActivate: [{ type: Output, args: ['activate',] }],
        onDeactivate: [{ type: Output, args: ['deactivate',] }],
        isActive: [{ type: HostBinding, args: ['class.active',] }, { type: Input }],
        isDisabled: [{ type: HostBinding, args: ['class.disabled',] }, { type: Input }],
        onClick: [{ type: HostListener, args: ['click',] }]
    };
    return FuiTabHeader;
}());
export { FuiTabHeader };
if (false) {
    /** @type {?} */
    FuiTabHeader.prototype.hasClasses;
    /** @type {?} */
    FuiTabHeader.prototype.id;
    /** @type {?} */
    FuiTabHeader.prototype.isActiveChange;
    /** @type {?} */
    FuiTabHeader.prototype.isActiveExternalChange;
    /** @type {?} */
    FuiTabHeader.prototype.onActivate;
    /** @type {?} */
    FuiTabHeader.prototype.onDeactivate;
    /**
     * @type {?}
     * @private
     */
    FuiTabHeader.prototype._isActive;
    /**
     * @type {?}
     * @private
     */
    FuiTabHeader.prototype._isDisabled;
}
