// ngx-fomantic-ui/esm5/modules/accordion/components/accordion.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, HostBinding, Input, QueryList } from '@angular/core';
import { FuiAccordionPanel } from './accordion-panel';
import { FuiAccordionService } from '../services/accordion.service';
var FuiAccordion = /** @class */ (function () {
    function FuiAccordion() {
        // Accordion service is unique to each set of panels.
        this._service = new FuiAccordionService();
        this.hasClasses = true;
    }
    Object.defineProperty(FuiAccordion.prototype, "closeOthers", {
        get: /**
         * @return {?}
         */
        function () {
            return this._service.closeOthers;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._service.closeOthers = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiAccordion.prototype, "transition", {
        set: /**
         * @param {?} transition
         * @return {?}
         */
        function (transition) {
            this._service.transition = transition;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiAccordion.prototype, "transitionDuration", {
        set: /**
         * @param {?} duration
         * @return {?}
         */
        function (duration) {
            this._service.transitionDuration = duration;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FuiAccordion.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.updatePanels();
        // Reconnect panels after they have updated.
        this._panels.changes.subscribe((/**
         * @return {?}
         */
        function () { return _this.updatePanels(); }));
    };
    /**
     * @return {?}
     */
    FuiAccordion.prototype.updatePanels = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._panels.forEach((/**
         * @param {?} p
         * @return {?}
         */
        function (p) { return _this._service.addPanel(p); }));
    };
    FuiAccordion.decorators = [
        { type: Component, args: [{
                    selector: 'fui-accordion',
                    template: "\n<ng-content></ng-content>\n",
                    styles: ["\n/* Fix for general styling issues */\n:host {\n    display: block;\n}\n\n/* Fix for styled border issue */\n:host.styled fui-accordion-panel:first-child .title {\n    border-top: none\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiAccordion.ctorParameters = function () { return []; };
    FuiAccordion.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.accordion',] }],
        _panels: [{ type: ContentChildren, args: [FuiAccordionPanel,] }],
        closeOthers: [{ type: Input }],
        transition: [{ type: Input }],
        transitionDuration: [{ type: Input }]
    };
    return FuiAccordion;
}());
export { FuiAccordion };
if (false) {
    /** @type {?} */
    FuiAccordion.prototype.hasClasses;
    /**
     * @type {?}
     * @protected
     */
    FuiAccordion.prototype._service;
    /**
     * @type {?}
     * @protected
     */
    FuiAccordion.prototype._panels;
}
