// ngx-fomantic-ui/esm5/modules/modal/classes/active-modal.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Helper class to support method chaining when calling `FuiModalService.open(...)`.
/**
 * @template T, U, V
 */
var 
// Helper class to support method chaining when calling `FuiModalService.open(...)`.
/**
 * @template T, U, V
 */
ActiveModal = /** @class */ (function () {
    function ActiveModal(instance, componentRef) {
        var _this = this;
        this._config = instance;
        this._componentRef = componentRef;
        // Automatically destroy the modal component when it has been dismissed.
        this.component.onDismiss.subscribe((/**
         * @return {?}
         */
        function () { return _this._componentRef.destroy(); }));
    }
    Object.defineProperty(ActiveModal.prototype, "component", {
        // Shorthand for direct access to the `FuiModal` instance.
        get: 
        // Shorthand for direct access to the `FuiModal` instance.
        /**
         * @return {?}
         */
        function () {
            return this._componentRef.instance;
        },
        enumerable: true,
        configurable: true
    });
    // Registers a callback for when `onApprove` is fired.
    // Registers a callback for when `onApprove` is fired.
    /**
     * @param {?} callback
     * @return {?}
     */
    ActiveModal.prototype.onApprove = 
    // Registers a callback for when `onApprove` is fired.
    /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        this.component.onApprove.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return callback(res); }));
        return this;
    };
    // Registers a callback for when `onDeny` is fired.
    // Registers a callback for when `onDeny` is fired.
    /**
     * @param {?} callback
     * @return {?}
     */
    ActiveModal.prototype.onDeny = 
    // Registers a callback for when `onDeny` is fired.
    /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        this.component.onDeny.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return callback(res); }));
        return this;
    };
    // Fires the approve event of the modal manually.
    // Fires the approve event of the modal manually.
    /**
     * @param {?} result
     * @return {?}
     */
    ActiveModal.prototype.approve = 
    // Fires the approve event of the modal manually.
    /**
     * @param {?} result
     * @return {?}
     */
    function (result) {
        this.component.approve(result);
    };
    // Fires the deny event of the modal manually.
    // Fires the deny event of the modal manually.
    /**
     * @param {?} result
     * @return {?}
     */
    ActiveModal.prototype.deny = 
    // Fires the deny event of the modal manually.
    /**
     * @param {?} result
     * @return {?}
     */
    function (result) {
        this.component.deny(result);
    };
    // Removes the modal component instantly, without transitions or firing any events.
    // Removes the modal component instantly, without transitions or firing any events.
    /**
     * @return {?}
     */
    ActiveModal.prototype.destroy = 
    // Removes the modal component instantly, without transitions or firing any events.
    /**
     * @return {?}
     */
    function () {
        this._componentRef.destroy();
    };
    return ActiveModal;
}());
// Helper class to support method chaining when calling `FuiModalService.open(...)`.
/**
 * @template T, U, V
 */
export { ActiveModal };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActiveModal.prototype._config;
    /**
     * @type {?}
     * @private
     */
    ActiveModal.prototype._componentRef;
}
