// ngx-fomantic-ui/esm5/modules/modal/classes/modal-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/** @type {?} */
export var ModalSize = {
    Mini: (/** @type {?} */ ('mini')),
    Tiny: (/** @type {?} */ ('tiny')),
    Small: (/** @type {?} */ ('small')),
    Normal: (/** @type {?} */ ('normal')),
    Large: (/** @type {?} */ ('large'))
};
// Stores a basic set of configuration options for a modal.
/**
 * @template T, U, V
 */
var 
// Stores a basic set of configuration options for a modal.
/**
 * @template T, U, V
 */
ModalConfig = /** @class */ (function () {
    function ModalConfig(context, isClosable) {
        if (context === void 0) { context = undefined; }
        if (isClosable === void 0) { isClosable = true; }
        // Initialise with default values.
        this.isClosable = isClosable;
        this.context = context;
        this.size = ModalSize.Normal;
        this.isFullScreen = false;
        this.isBasic = false;
        this.isInverted = false;
        this.isCentered = true;
        this.mustScroll = false;
        this.transition = 'scale';
        this.transitionDuration = 500;
    }
    return ModalConfig;
}());
// Stores a basic set of configuration options for a modal.
/**
 * @template T, U, V
 */
export { ModalConfig };
if (false) {
    /** @type {?} */
    ModalConfig.prototype.isClosable;
    /** @type {?} */
    ModalConfig.prototype.closeResult;
    /** @type {?} */
    ModalConfig.prototype.context;
    /** @type {?} */
    ModalConfig.prototype.size;
    /** @type {?} */
    ModalConfig.prototype.isFullScreen;
    /** @type {?} */
    ModalConfig.prototype.isBasic;
    /** @type {?} */
    ModalConfig.prototype.isInverted;
    /** @type {?} */
    ModalConfig.prototype.isCentered;
    /** @type {?} */
    ModalConfig.prototype.mustScroll;
    /** @type {?} */
    ModalConfig.prototype.transition;
    /** @type {?} */
    ModalConfig.prototype.transitionDuration;
}
// Used when creating a modal from a `TemplateRef`.
/**
 * @template T, U, V
 */
var 
// Used when creating a modal from a `TemplateRef`.
/**
 * @template T, U, V
 */
TemplateModalConfig = /** @class */ (function (_super) {
    tslib_1.__extends(TemplateModalConfig, _super);
    function TemplateModalConfig(template, context, isClosable) {
        if (context === void 0) { context = undefined; }
        if (isClosable === void 0) { isClosable = true; }
        var _this = _super.call(this, context, isClosable) || this;
        _this.template = template;
        return _this;
    }
    return TemplateModalConfig;
}(ModalConfig));
// Used when creating a modal from a `TemplateRef`.
/**
 * @template T, U, V
 */
export { TemplateModalConfig };
if (false) {
    /** @type {?} */
    TemplateModalConfig.prototype.template;
}
// Used when creating a modal from an existing component.
/**
 * @template T, U, V
 */
var 
// Used when creating a modal from an existing component.
/**
 * @template T, U, V
 */
ComponentModalConfig = /** @class */ (function (_super) {
    tslib_1.__extends(ComponentModalConfig, _super);
    function ComponentModalConfig(component, context, isClosable) {
        if (context === void 0) { context = undefined; }
        if (isClosable === void 0) { isClosable = true; }
        var _this = _super.call(this, context, isClosable) || this;
        _this.component = component;
        return _this;
    }
    return ComponentModalConfig;
}(ModalConfig));
// Used when creating a modal from an existing component.
/**
 * @template T, U, V
 */
export { ComponentModalConfig };
if (false) {
    /** @type {?} */
    ComponentModalConfig.prototype.component;
}
