// ngx-fomantic-ui/esm5/modules/transition/classes/transition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var TransitionDirection = {
    In: 0,
    Out: 1,
    Either: 2,
    Static: 3,
};
export { TransitionDirection };
TransitionDirection[TransitionDirection.In] = 'In';
TransitionDirection[TransitionDirection.Out] = 'Out';
TransitionDirection[TransitionDirection.Either] = 'Either';
TransitionDirection[TransitionDirection.Static] = 'Static';
var Transition = /** @class */ (function () {
    function Transition(name, duration, direction, onComplete) {
        if (duration === void 0) { duration = 250; }
        if (direction === void 0) { direction = TransitionDirection.Either; }
        if (onComplete === void 0) { onComplete = (/**
         * @return {?}
         */
        function () {
        }); }
        this.type = name;
        // We set a minimum duration of 1ms, to give the appearance of an immediate transition
        // whilst allowing positioning calculations to happen without a visible flicker.
        this.duration = Math.max(duration, 1);
        this.direction = direction;
        this.classes = this.type.split(' ');
        this.onComplete = onComplete;
    }
    Object.defineProperty(Transition.prototype, "directionClass", {
        // Converts TransitionDirection to class name.
        get: 
        // Converts TransitionDirection to class name.
        /**
         * @return {?}
         */
        function () {
            switch (this.direction) {
                case TransitionDirection.In:
                    return 'in';
                case TransitionDirection.Out:
                    return 'out';
            }
            return '';
        },
        enumerable: true,
        configurable: true
    });
    return Transition;
}());
export { Transition };
if (false) {
    /** @type {?} */
    Transition.prototype.type;
    /** @type {?} */
    Transition.prototype.duration;
    /** @type {?} */
    Transition.prototype.direction;
    /** @type {?} */
    Transition.prototype.classes;
    /** @type {?} */
    Transition.prototype.onComplete;
}
