// ngx-fomantic-ui/esm2015/modules/datepicker/directives/calendar-item.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, EventEmitter, HostBinding, HostListener, Input } from '@angular/core';
export class CalendarItem {
    /**
     * @param {?} date
     */
    constructor(date) {
        this.date = date;
    }
}
if (false) {
    /** @type {?} */
    CalendarItem.prototype.date;
    /** @type {?} */
    CalendarItem.prototype.humanReadable;
    /** @type {?} */
    CalendarItem.prototype.isDisabled;
    /** @type {?} */
    CalendarItem.prototype.isActive;
    /** @type {?} */
    CalendarItem.prototype.isOutsideRange;
    /** @type {?} */
    CalendarItem.prototype.isToday;
    /** @type {?} */
    CalendarItem.prototype.isSelectable;
}
export class FuiCalendarItem {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.hasFocus = false;
        this.onFocussed = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get isSelectable() {
        return this.item.isSelectable;
    }
    /**
     * @return {?}
     */
    get isActive() {
        return this.item.isActive;
    }
    /**
     * @return {?}
     */
    get isToday() {
        return this.item.isToday;
    }
    /**
     * @return {?}
     */
    onMouseMove() {
        if (!this.hasFocus) {
            this.hasFocus = true;
            this.onFocussed.emit(this.hasFocus);
        }
    }
    /**
     * @return {?}
     */
    onMouseLeave() {
        this.hasFocus = false;
        this.onFocussed.emit(this.hasFocus);
    }
}
FuiCalendarItem.decorators = [
    { type: Directive, args: [{
                selector: '[calendarItem]'
            },] }
];
/** @nocollapse */
FuiCalendarItem.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
FuiCalendarItem.propDecorators = {
    item: [{ type: Input, args: ['calendarItem',] }],
    hasFocus: [{ type: HostBinding, args: ['class.focus',] }],
    isSelectable: [{ type: HostBinding, args: ['class.disabled',] }],
    isActive: [{ type: HostBinding, args: ['class.active',] }],
    isToday: [{ type: HostBinding, args: ['class.today',] }],
    onMouseMove: [{ type: HostListener, args: ['mousemove',] }],
    onMouseLeave: [{ type: HostListener, args: ['mouseleave',] }]
};
if (false) {
    /** @type {?} */
    FuiCalendarItem.prototype.item;
    /** @type {?} */
    FuiCalendarItem.prototype.hasFocus;
    /** @type {?} */
    FuiCalendarItem.prototype.onFocussed;
    /** @type {?} */
    FuiCalendarItem.prototype.changeDetector;
}
