// ngx-fomantic-ui/esm2015/modules/datepicker/components/calendar-view-title.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { CalendarRangeService } from '../services/calendar-range.service';
export class FuiCalendarViewTitle {
    constructor() {
        this.onZoomOut = new EventEmitter();
    }
}
FuiCalendarViewTitle.decorators = [
    { type: Component, args: [{
                selector: 'fui-calendar-view-title',
                template: `
<span class="title link" (click)="onZoomOut.emit()">
    <ng-content></ng-content>
</span>
<span class="prev link" [class.disabled]="!ranges?.canMovePrevious" (click)="ranges?.movePrevious()">
    <i class="chevron left icon"></i>
</span>
<span class="next link" [class.disabled]="!ranges?.canMoveNext" (click)="ranges?.moveNext()">
    <i class="chevron right icon"></i>
</span>
`,
                styles: [`
.title.link {
    display: inline-block;
    margin-left: 2rem;
    margin-right: 2rem;
}
`]
            }] }
];
/** @nocollapse */
FuiCalendarViewTitle.ctorParameters = () => [];
FuiCalendarViewTitle.propDecorators = {
    ranges: [{ type: Input }],
    onZoomOut: [{ type: Output, args: ['zoomOut',] }]
};
if (false) {
    /** @type {?} */
    FuiCalendarViewTitle.prototype.ranges;
    /** @type {?} */
    FuiCalendarViewTitle.prototype.onZoomOut;
}
