// ngx-fomantic-ui/esm2015/modules/datepicker/components/datepicker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, HostListener } from '@angular/core';
import { CalendarService } from './../services/calendar.service';
import { DatetimeConfig } from '../classes/calendar-config';
import { FuiLocalizationService } from '../../../behaviors/localization/internal';
/** @type {?} */
export const DatepickerMode = {
    Year: (/** @type {?} */ ('year')),
    Month: (/** @type {?} */ ('month')),
    Date: (/** @type {?} */ ('date')),
    Datetime: (/** @type {?} */ ('datetime')),
    Time: (/** @type {?} */ ('time'))
};
export class FuiDatepicker {
    /**
     * @param {?} localizationService
     */
    constructor(localizationService) {
        this.service = new CalendarService(new DatetimeConfig(), localizationService.get().datepicker);
        this.hasClasses = true;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onMouseDown(e) {
        e.preventDefault();
    }
}
FuiDatepicker.decorators = [
    { type: Component, args: [{
                selector: 'fui-datepicker',
                template: `
<ng-container [ngSwitch]="service.currentView">
    <fui-calendar-year-view [service]="service" *ngSwitchCase="0"></fui-calendar-year-view>
    <fui-calendar-month-view [service]="service" *ngSwitchCase="1"></fui-calendar-month-view>
    <fui-calendar-date-view [service]="service" *ngSwitchCase="2"></fui-calendar-date-view>
    <fui-calendar-hour-view [service]="service" *ngSwitchCase="3"></fui-calendar-hour-view>
    <fui-calendar-minute-view [service]="service" *ngSwitchCase="4"></fui-calendar-minute-view>
</ng-container>
`,
                styles: [`
:host {
    user-select: none;
}
`]
            }] }
];
/** @nocollapse */
FuiDatepicker.ctorParameters = () => [
    { type: FuiLocalizationService }
];
FuiDatepicker.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.active',] }, { type: HostBinding, args: ['class.calendar',] }],
    onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    FuiDatepicker.prototype.hasClasses;
    /** @type {?} */
    FuiDatepicker.prototype.service;
}
