// ngx-fomantic-ui/esm5/modules/toast/classes/toast-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var FuiToastContainer = /** @class */ (function () {
    function FuiToastContainer() {
        this.toastData = [];
        this.timeoutIds = {};
        this.maxId = 0;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    FuiToastContainer.prototype.addToast = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        data.id = data.id || ++this.maxId;
        data.showProgress = data.showProgress || 'bottom';
        data.class = data.class || 'info';
        data.displayTime = data.displayTime || 5000;
        this.toastData.unshift(data);
        this.timeoutIds[data.id] = window.setTimeout((/**
         * @return {?}
         */
        function () { return _this.closeToast(data.id); }), data.displayTime);
    };
    /**
     * @param {?} toastId
     * @return {?}
     */
    FuiToastContainer.prototype.closeToast = /**
     * @param {?} toastId
     * @return {?}
     */
    function (toastId) {
        /** @type {?} */
        var x = this.toastData.findIndex((/**
         * @param {?} t
         * @return {?}
         */
        function (t) { return t.id === toastId; }));
        if (x > -1) {
            /** @type {?} */
            var id = this.toastData[x].id;
            window.clearTimeout(this.timeoutIds[id]);
            delete this.timeoutIds[id];
            this.toastData.splice(x, 1);
        }
    };
    FuiToastContainer.decorators = [
        { type: Component, args: [{
                    selector: 'fui-toast-container',
                    exportAs: 'fuiToastContainer',
                    template: "\n    <fui-toast *ngFor=\"let toast of toastData\"\n               [id]=\"toast.id\"\n               [title]=\"toast.title\"\n               [message]=\"toast.message\"\n               [class]=\"toast.class\"\n               [showProgress]=\"toast.showProgress\"\n               [displayTime]=\"toast.displayTime\"\n               [showIcon]=\"toast.showIcon\"\n               [closeIcon]=\"toast.closeIcon\"\n               [progressUp]=\"toast.progressUp\"\n               [className]=\"toast.className\"\n               (close)=\"closeToast(toast.id)\"></fui-toast>"
                }] }
    ];
    return FuiToastContainer;
}());
export { FuiToastContainer };
if (false) {
    /** @type {?} */
    FuiToastContainer.prototype.toastData;
    /** @type {?} */
    FuiToastContainer.prototype.timeoutIds;
    /**
     * @type {?}
     * @private
     */
    FuiToastContainer.prototype.maxId;
}
