// ngx-fomantic-ui/esm2015/modules/checkbox/directives/radio-manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, Directive, ElementRef, QueryList } from '@angular/core';
import { FuiRadio } from '../components/radio';
import { Util } from '../../../misc/util/internal';
/**
 * @template T
 */
export class FuiRadioManager {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
        this.isNested = false;
        this._radioSubs = [];
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.updateNesting();
        this._subManagers.changes.subscribe((/**
         * @return {?}
         */
        () => this.updateNesting()));
        this.updateRadios();
        this._renderedRadios.changes.subscribe((/**
         * @return {?}
         */
        () => this.updateRadios()));
    }
    /**
     * @private
     * @return {?}
     */
    updateNesting() {
        this._subManagers
            .filter((/**
         * @param {?} m
         * @return {?}
         */
        m => m !== this))
            .forEach((/**
         * @param {?} m
         * @return {?}
         */
        m => m.isNested = true));
    }
    /**
     * @private
     * @return {?}
     */
    updateRadios() {
        this._radioSubs.forEach((/**
         * @param {?} s
         * @return {?}
         */
        s => s.unsubscribe()));
        this._radioSubs = [];
        /** @type {?} */
        const groups = Util.Array.groupBy(this._renderedRadios.toArray(), 'name');
        Object
            .keys(groups)
            .map((/**
         * @param {?} k
         * @return {?}
         */
        k => groups[k]))
            .forEach((/**
         * @param {?} g
         * @return {?}
         */
        g => g
            .forEach((/**
         * @param {?} r
         * @return {?}
         */
        r => this._radioSubs
            .push(r.onCurrentValueChange
            .subscribe((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            if (!this.isNested) {
                g.forEach((/**
                 * @param {?} radio
                 * @return {?}
                 */
                radio => radio.writeValue(v)));
            }
        })))))));
    }
}
FuiRadioManager.decorators = [
    { type: Directive, args: [{
                selector: 'form:not([ngForm]):not([[ngForm]]),ngForm,[ngForm]'
            },] }
];
/** @nocollapse */
FuiRadioManager.ctorParameters = () => [
    { type: ElementRef }
];
FuiRadioManager.propDecorators = {
    _subManagers: [{ type: ContentChildren, args: [FuiRadioManager, { descendants: true },] }],
    _renderedRadios: [{ type: ContentChildren, args: [FuiRadio, { descendants: true },] }]
};
if (false) {
    /** @type {?} */
    FuiRadioManager.prototype.isNested;
    /**
     * @type {?}
     * @private
     */
    FuiRadioManager.prototype._subManagers;
    /**
     * @type {?}
     * @private
     */
    FuiRadioManager.prototype._renderedRadios;
    /**
     * @type {?}
     * @private
     */
    FuiRadioManager.prototype._radioSubs;
    /** @type {?} */
    FuiRadioManager.prototype.element;
}
