// ngx-fomantic-ui/esm2015/modules/rating/components/rating.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Directive, EventEmitter, HostBinding, HostListener, Input, Output } from '@angular/core';
import { CustomValueAccessor, customValueAccessorFactory } from '../../../misc/util/internal';
export class FuiRating {
    constructor() {
        this.hoveredIndex = -1;
        this.value = 0;
        this.valueChange = new EventEmitter();
        this.type = 'star';
        this.maximum = 5;
        this.isReadonly = false;
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get icons() {
        // tslint:disable-next-line:prefer-literal
        return new Array(this.maximum);
    }
    /**
     * @return {?}
     */
    get type() {
        return this._type;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @return {?}
     */
    get maximum() {
        return this._maximum;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maximum(value) {
        this._maximum = +value;
    }
    /**
     * @param {?} i
     * @return {?}
     */
    onClick(i) {
        if (!this.isReadonly) {
            this.value = i + 1;
            this.valueChange.emit(this.value);
        }
    }
    /**
     * @param {?} i
     * @return {?}
     */
    onMouseover(i) {
        this.hoveredIndex = i;
    }
    /**
     * @return {?}
     */
    onMouseout() {
        this.hoveredIndex = -1;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.value = value;
    }
}
FuiRating.decorators = [
    { type: Component, args: [{
                selector: 'fui-rating',
                // tslint:disable-next-line
                template: `<i *ngFor="let icon of icons; let i = index" class="icon {{type}}" (mouseover)="onMouseover(i)" (click)="onClick(i)" [class.selected]="hoveredIndex >= i && !isReadonly" [class.active]="value > i"></i>`,
                styles: [`
:host.read-only .icon {
    cursor: auto
}
`]
            }] }
];
/** @nocollapse */
FuiRating.ctorParameters = () => [];
FuiRating.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.rating',] }],
    valueChange: [{ type: Output }],
    isReadonly: [{ type: HostBinding, args: ['class.read-only',] }, { type: Input }],
    type: [{ type: Input }],
    maximum: [{ type: Input }],
    onMouseout: [{ type: HostListener, args: ['mouseout',] }]
};
if (false) {
    /** @type {?} */
    FuiRating.prototype.hasClasses;
    /** @type {?} */
    FuiRating.prototype.value;
    /** @type {?} */
    FuiRating.prototype.valueChange;
    /** @type {?} */
    FuiRating.prototype.isReadonly;
    /** @type {?} */
    FuiRating.prototype.hoveredIndex;
    /**
     * @type {?}
     * @private
     */
    FuiRating.prototype._type;
    /**
     * @type {?}
     * @private
     */
    FuiRating.prototype._maximum;
}
export class FuiRatingValueAccessor extends CustomValueAccessor {
    /**
     * @param {?} host
     */
    constructor(host) {
        super(host);
    }
}
FuiRatingValueAccessor.decorators = [
    { type: Directive, args: [{
                selector: 'fui-rating',
                host: { '(valueChange)': 'onChange($event)' },
                providers: [customValueAccessorFactory(FuiRatingValueAccessor)]
            },] }
];
/** @nocollapse */
FuiRatingValueAccessor.ctorParameters = () => [
    { type: FuiRating }
];
