// ngx-fomantic-ui/esm2015/modules/toast/classes/toast-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
export class FuiToastContainer {
    constructor() {
        this.toastData = [];
        this.timeoutIds = {};
        this.maxId = 0;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    addToast(data) {
        data.id = data.id || ++this.maxId;
        data.showProgress = data.showProgress || 'bottom';
        data.class = data.class || 'info';
        data.displayTime = data.displayTime || 5000;
        this.toastData.unshift(data);
        this.timeoutIds[data.id] = window.setTimeout((/**
         * @return {?}
         */
        () => this.closeToast(data.id)), data.displayTime);
    }
    /**
     * @param {?} toastId
     * @return {?}
     */
    closeToast(toastId) {
        /** @type {?} */
        const x = this.toastData.findIndex((/**
         * @param {?} t
         * @return {?}
         */
        t => t.id === toastId));
        if (x > -1) {
            /** @type {?} */
            const id = this.toastData[x].id;
            window.clearTimeout(this.timeoutIds[id]);
            delete this.timeoutIds[id];
            this.toastData.splice(x, 1);
        }
    }
}
FuiToastContainer.decorators = [
    { type: Component, args: [{
                selector: 'fui-toast-container',
                exportAs: 'fuiToastContainer',
                template: `
    <fui-toast *ngFor="let toast of toastData"
               [id]="toast.id"
               [title]="toast.title"
               [message]="toast.message"
               [class]="toast.class"
               [showProgress]="toast.showProgress"
               [displayTime]="toast.displayTime"
               [showIcon]="toast.showIcon"
               [closeIcon]="toast.closeIcon"
               [progressUp]="toast.progressUp"
               [className]="toast.className"
               (close)="closeToast(toast.id)"></fui-toast>`
            }] }
];
if (false) {
    /** @type {?} */
    FuiToastContainer.prototype.toastData;
    /** @type {?} */
    FuiToastContainer.prototype.timeoutIds;
    /**
     * @type {?}
     * @private
     */
    FuiToastContainer.prototype.maxId;
}
