// ngx-fomantic-ui/esm5/modules/checkbox/directives/radio-manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, Directive, ElementRef, QueryList } from '@angular/core';
import { FuiRadio } from '../components/radio';
import { Util } from '../../../misc/util/internal';
/**
 * @template T
 */
var FuiRadioManager = /** @class */ (function () {
    function FuiRadioManager(element) {
        this.element = element;
        this.isNested = false;
        this._radioSubs = [];
    }
    /**
     * @return {?}
     */
    FuiRadioManager.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.updateNesting();
        this._subManagers.changes.subscribe((/**
         * @return {?}
         */
        function () { return _this.updateNesting(); }));
        this.updateRadios();
        this._renderedRadios.changes.subscribe((/**
         * @return {?}
         */
        function () { return _this.updateRadios(); }));
    };
    /**
     * @private
     * @return {?}
     */
    FuiRadioManager.prototype.updateNesting = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this._subManagers
            .filter((/**
         * @param {?} m
         * @return {?}
         */
        function (m) { return m !== _this; }))
            .forEach((/**
         * @param {?} m
         * @return {?}
         */
        function (m) { return m.isNested = true; }));
    };
    /**
     * @private
     * @return {?}
     */
    FuiRadioManager.prototype.updateRadios = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this._radioSubs.forEach((/**
         * @param {?} s
         * @return {?}
         */
        function (s) { return s.unsubscribe(); }));
        this._radioSubs = [];
        /** @type {?} */
        var groups = Util.Array.groupBy(this._renderedRadios.toArray(), 'name');
        Object
            .keys(groups)
            .map((/**
         * @param {?} k
         * @return {?}
         */
        function (k) { return groups[k]; }))
            .forEach((/**
         * @param {?} g
         * @return {?}
         */
        function (g) { return g
            .forEach((/**
         * @param {?} r
         * @return {?}
         */
        function (r) { return _this._radioSubs
            .push(r.onCurrentValueChange
            .subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (!_this.isNested) {
                g.forEach((/**
                 * @param {?} radio
                 * @return {?}
                 */
                function (radio) { return radio.writeValue(v); }));
            }
        }))); })); }));
    };
    FuiRadioManager.decorators = [
        { type: Directive, args: [{
                    selector: 'form:not([ngForm]):not([[ngForm]]),ngForm,[ngForm]'
                },] }
    ];
    /** @nocollapse */
    FuiRadioManager.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FuiRadioManager.propDecorators = {
        _subManagers: [{ type: ContentChildren, args: [FuiRadioManager, { descendants: true },] }],
        _renderedRadios: [{ type: ContentChildren, args: [FuiRadio, { descendants: true },] }]
    };
    return FuiRadioManager;
}());
export { FuiRadioManager };
if (false) {
    /** @type {?} */
    FuiRadioManager.prototype.isNested;
    /**
     * @type {?}
     * @private
     */
    FuiRadioManager.prototype._subManagers;
    /**
     * @type {?}
     * @private
     */
    FuiRadioManager.prototype._renderedRadios;
    /**
     * @type {?}
     * @private
     */
    FuiRadioManager.prototype._radioSubs;
    /** @type {?} */
    FuiRadioManager.prototype.element;
}
