// ngx-fomantic-ui/esm2015/modules/transition/directives/transition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, HostBinding, Input, Renderer2 } from '@angular/core';
import { TransitionController } from '../classes/transition-controller';
export class FuiTransition {
    /**
     * @param {?} _renderer
     * @param {?} _element
     * @param {?} _changeDetector
     */
    constructor(_renderer, _element, _changeDetector) {
        this._renderer = _renderer;
        this._element = _element;
        this._changeDetector = _changeDetector;
        this.transitionClass = true;
    }
    /**
     * @param {?} tC
     * @return {?}
     */
    set fuiTransition(tC) {
        // Set the transition controller (e.g. '<div [fuiTransition]="transitionController"></div>').
        this.setTransitionController(tC);
    }
    /**
     * @return {?}
     */
    get isVisible() {
        if (this._controller) {
            return this._controller.isVisible;
        }
        return false;
    }
    /**
     * @return {?}
     */
    get isHidden() {
        if (this._controller) {
            return this._controller.isHidden;
        }
        return false;
    }
    // Initialises the controller with the injected renderer and elementRef.
    /**
     * @param {?} transitionController
     * @return {?}
     */
    setTransitionController(transitionController) {
        this._controller = transitionController;
        this._controller.registerRenderer(this._renderer);
        this._controller.registerElement(this._element.nativeElement);
        this._controller.registerChangeDetector(this._changeDetector);
    }
}
FuiTransition.decorators = [
    { type: Directive, args: [{
                selector: '[fuiTransition]',
                exportAs: 'transition'
            },] }
];
/** @nocollapse */
FuiTransition.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
FuiTransition.propDecorators = {
    transitionClass: [{ type: HostBinding, args: ['class.transition',] }],
    fuiTransition: [{ type: Input }],
    isVisible: [{ type: HostBinding, args: ['class.visible',] }],
    isHidden: [{ type: HostBinding, args: ['class.hidden',] }]
};
if (false) {
    /** @type {?} */
    FuiTransition.prototype.transitionClass;
    /**
     * @type {?}
     * @private
     */
    FuiTransition.prototype._controller;
    /**
     * @type {?}
     * @protected
     */
    FuiTransition.prototype._renderer;
    /**
     * @type {?}
     * @protected
     */
    FuiTransition.prototype._element;
    /**
     * @type {?}
     * @private
     */
    FuiTransition.prototype._changeDetector;
}
