// ngx-fomantic-ui/esm5/modules/progress/components/progress.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Input } from '@angular/core';
var FuiProgress = /** @class */ (function () {
    function FuiProgress() {
        this.value = 0;
        this.maximum = 100;
        this.precision = 0;
        this._overrideSuccess = false;
        this.autoSuccess = true;
        this.showProgress = true;
        this.hasClasses = true;
    }
    Object.defineProperty(FuiProgress.prototype, "reachedMaximum", {
        get: /**
         * @return {?}
         */
        function () {
            return this._overrideSuccess || ((this.value >= this.maximum) && this.autoSuccess);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiProgress.prototype, "percentage", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var boundedValue = Math.min(Math.max(this.value, 0), this.maximum);
            /** @type {?} */
            var percentage = (boundedValue / this.maximum) * 100;
            return percentage.toFixed(this.precision);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiProgress.prototype, "classValue", {
        set: /**
         * @param {?} classes
         * @return {?}
         */
        function (classes) {
            if (classes.includes('attached') || classes.includes('tiny')) {
                this.showProgress = false;
            }
            if (classes.includes('success')) {
                this._overrideSuccess = true;
            }
            this.transition = 'ease';
            this.transitionDuration = 350;
            this.canCompletelyEmpty = false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiProgress.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            // Convert value from string to number where necessary.
            /** @type {?} */
            var converted = +value;
            if (Number.isNaN(converted)) {
                return;
            }
            this._value = converted;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiProgress.prototype, "maximum", {
        get: /**
         * @return {?}
         */
        function () {
            return this._maximum;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            // Convert value from string to number where necessary.
            /** @type {?} */
            var converted = +value;
            if (Number.isNaN(converted)) {
                return;
            }
            this._maximum = converted;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiProgress.prototype, "precision", {
        get: /**
         * @return {?}
         */
        function () {
            return this._precision;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            // Convert value from string to number where necessary.
            /** @type {?} */
            var converted = +value;
            if (Number.isNaN(converted)) {
                return;
            }
            this._precision = Math.min(Math.max(converted, 0), 20);
        },
        enumerable: true,
        configurable: true
    });
    FuiProgress.decorators = [
        { type: Component, args: [{
                    selector: 'fui-progress',
                    template: "\n<div class=\"bar\"\n    [style.width.%]=\"percentage\"\n    [style.minWidth]=\"canCompletelyEmpty ? 0 : null\"\n    [style.transitionTimingFunction]=\"transition\"\n    [style.transitionDuration.ms]=\"transitionDuration\">\n    <div class=\"progress\" *ngIf=\"showProgress\">{{ percentage }}%</div>\n</div>\n<div class=\"label\">\n    <ng-content></ng-content>\n</div>\n",
                    styles: ["\n.bar {\n    transition-duration: 300ms !important;\n    z-index: 1;\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiProgress.ctorParameters = function () { return []; };
    FuiProgress.propDecorators = {
        transition: [{ type: Input }],
        transitionDuration: [{ type: Input }],
        canCompletelyEmpty: [{ type: Input }],
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.progress',] }],
        autoSuccess: [{ type: Input }],
        showProgress: [{ type: Input }],
        reachedMaximum: [{ type: HostBinding, args: ['class.success',] }],
        percentage: [{ type: HostBinding, args: ['attr.data-percent',] }],
        classValue: [{ type: Input, args: ['class',] }],
        value: [{ type: Input }],
        maximum: [{ type: Input }],
        precision: [{ type: Input }]
    };
    return FuiProgress;
}());
export { FuiProgress };
if (false) {
    /** @type {?} */
    FuiProgress.prototype.transition;
    /** @type {?} */
    FuiProgress.prototype.transitionDuration;
    /** @type {?} */
    FuiProgress.prototype.canCompletelyEmpty;
    /** @type {?} */
    FuiProgress.prototype.hasClasses;
    /** @type {?} */
    FuiProgress.prototype.autoSuccess;
    /** @type {?} */
    FuiProgress.prototype.showProgress;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._overrideSuccess;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._value;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._maximum;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._precision;
}
