// ngx-fomantic-ui/esm5/modules/popup/classes/popup-template-controller.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { TemplateRef } from '@angular/core';
import { FuiPopupController } from './popup-controller';
import { PopupConfig } from './popup-config';
/** @type {?} */
var templateRef = TemplateRef;
/**
 * @record
 * @template T
 */
export function ITemplatePopupContext() { }
if (false) {
    /** @type {?|undefined} */
    ITemplatePopupContext.prototype.context;
}
/**
 * @record
 * @template T
 */
export function ITemplatePopupConfig() { }
if (false) {
    /** @type {?|undefined} */
    ITemplatePopupConfig.prototype.template;
    /** @type {?|undefined} */
    ITemplatePopupConfig.prototype.context;
}
/**
 * @template T
 */
var /**
 * @template T
 */
TemplatePopupConfig = /** @class */ (function (_super) {
    tslib_1.__extends(TemplatePopupConfig, _super);
    function TemplatePopupConfig() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return TemplatePopupConfig;
}(PopupConfig));
/**
 * @template T
 */
export { TemplatePopupConfig };
if (false) {
    /** @type {?} */
    TemplatePopupConfig.prototype.template;
    /** @type {?} */
    TemplatePopupConfig.prototype.context;
}
/**
 * @template T
 */
var /**
 * @template T
 */
FuiPopupTemplateController = /** @class */ (function (_super) {
    tslib_1.__extends(FuiPopupTemplateController, _super);
    function FuiPopupTemplateController(renderer, element, componentFactory, config) {
        return _super.call(this, renderer, element, componentFactory, config) || this;
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    FuiPopupTemplateController.prototype.configure = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        _super.prototype.configure.call(this, config);
        if (config) {
            this.template = config.template;
            this.context = config.context;
        }
    };
    /**
     * @return {?}
     */
    FuiPopupTemplateController.prototype.open = /**
     * @return {?}
     */
    function () {
        // If there is a template, inject it into the view.
        if (this.template) {
            this.popup.templateSibling.clear();
            this._componentFactory.createView(this.popup.templateSibling, this.template, {
                $implicit: this.popup,
                context: this.context
            });
        }
        _super.prototype.open.call(this);
    };
    return FuiPopupTemplateController;
}(FuiPopupController));
/**
 * @template T
 */
export { FuiPopupTemplateController };
if (false) {
    /** @type {?} */
    FuiPopupTemplateController.prototype.template;
    /** @type {?} */
    FuiPopupTemplateController.prototype.context;
}
