// ngx-fomantic-ui/esm2015/modules/sidebar/components/sidebar.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostBinding, Input, Output, Renderer2 } from '@angular/core';
import { SidebarDirection, SidebarService, SidebarTransition } from '../services/sidebar.service';
export class FuiSidebar {
    /**
     * @param {?} _renderer
     * @param {?} _element
     */
    constructor(_renderer, _element) {
        this._renderer = _renderer;
        this._element = _element;
        this.service = new SidebarService();
        // We set the default here as well to force the classes to update.
        this.transition = SidebarTransition.Uncover;
        this.direction = SidebarDirection.Left;
        setTimeout((/**
         * @return {?}
         */
        () => this.updateDimensions()));
        this.service.isVisibleChange.subscribe((/**
         * @return {?}
         */
        () => this.updateDimensions()));
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get transition() {
        return this.service.transition;
    }
    /**
     * @param {?} transition
     * @return {?}
     */
    set transition(transition) {
        this.service.transition.split(' ').forEach((/**
         * @param {?} c
         * @return {?}
         */
        c => this.setClass(c, false)));
        this.service.transition = transition;
        this.service.transition.split(' ').forEach((/**
         * @param {?} c
         * @return {?}
         */
        c => this.setClass(c, true)));
    }
    /**
     * @return {?}
     */
    get direction() {
        return this.service.direction;
    }
    /**
     * @param {?} direction
     * @return {?}
     */
    set direction(direction) {
        this.setClass(this.service.direction, false);
        this.service.direction = direction;
        this.setClass(this.service.direction, true);
    }
    /**
     * @return {?}
     */
    get isVisible() {
        return this.service.isVisible;
    }
    /**
     * @param {?} isVisible
     * @return {?}
     */
    set isVisible(isVisible) {
        this.service.setVisibleState(isVisible);
    }
    /**
     * @return {?}
     */
    get isVisibleChange() {
        return this.service.isVisibleChange;
    }
    /**
     * @return {?}
     */
    get isAnimating() {
        return this.service.isAnimating;
    }
    /**
     * @return {?}
     */
    open() {
        this.service.setVisibleState(true);
    }
    /**
     * @return {?}
     */
    close() {
        this.service.setVisibleState(false);
    }
    /**
     * @return {?}
     */
    toggle() {
        this.service.toggleVisibleState();
    }
    /**
     * @private
     * @return {?}
     */
    updateDimensions() {
        this.service.width = this._element.nativeElement.offsetWidth;
        this.service.height = this._element.nativeElement.offsetHeight;
    }
    /**
     * @private
     * @param {?} className
     * @param {?=} isAdd
     * @return {?}
     */
    setClass(className, isAdd = true) {
        if (isAdd) {
            this._renderer.addClass(this._element.nativeElement, className);
        }
        else {
            this._renderer.removeClass(this._element.nativeElement, className);
        }
    }
}
FuiSidebar.decorators = [
    { type: Component, args: [{
                selector: 'fui-sidebar',
                template: `<ng-content></ng-content>`
            }] }
];
/** @nocollapse */
FuiSidebar.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
FuiSidebar.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.sidebar',] }, { type: HostBinding, args: ['class.menu',] }],
    transition: [{ type: Input }],
    direction: [{ type: Input }],
    isVisible: [{ type: HostBinding, args: ['class.visible',] }, { type: Input }],
    isVisibleChange: [{ type: Output }],
    isAnimating: [{ type: HostBinding, args: ['class.animating',] }]
};
if (false) {
    /** @type {?} */
    FuiSidebar.prototype.service;
    /** @type {?} */
    FuiSidebar.prototype.hasClasses;
    /**
     * @type {?}
     * @private
     */
    FuiSidebar.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FuiSidebar.prototype._element;
}
