// ngx-fomantic-ui/esm2015/modules/sidebar/components/sidebar-sibling.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostBinding, HostListener, Input, Renderer2 } from '@angular/core';
import { SidebarTransition } from '../services/sidebar.service';
export class FuiSidebarSibling {
    /**
     * @param {?} _renderer
     * @param {?} _element
     */
    constructor(_renderer, _element) {
        this._renderer = _renderer;
        this._element = _element;
        this.canCloseSidebar = true;
        this.isDimmedWhenVisible = false;
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get isVisible() {
        if (!this.service) {
            return false;
        }
        return this.service.isVisible;
    }
    /**
     * @return {?}
     */
    get isDimmed() {
        if (!this.service) {
            return false;
        }
        return this.service.isVisible && this.isDimmedWhenVisible;
    }
    /**
     * @return {?}
     */
    get service() {
        return this._service;
    }
    /**
     * @param {?} service
     * @return {?}
     */
    set service(service) {
        this._service = service;
        setTimeout((/**
         * @return {?}
         */
        () => this.updateTransform()));
        this._service.isVisibleChange.subscribe((/**
         * @return {?}
         */
        () => this.updateTransform()));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        if (this.canCloseSidebar && this.service.isVisible && !this.service.wasJustOpened) {
            this.service.setVisibleState(false);
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateTransform() {
        this._renderer.removeStyle(this._element.nativeElement, 'transform');
        this._renderer.removeStyle(this._element.nativeElement, '-webkit-transform');
        if (this.service.isVisible &&
            this.service.transition !== SidebarTransition.Overlay &&
            this.service.transition !== SidebarTransition.ScaleDown) {
            /** @type {?} */
            const translate = `translate3d(${this.service.width}px, ${this.service.height}px, 0)`;
            this._renderer.setStyle(this._element.nativeElement, 'transform', translate);
            this._renderer.setStyle(this._element.nativeElement, '-webkit-transform', translate);
        }
    }
}
FuiSidebarSibling.decorators = [
    { type: Component, args: [{
                selector: 'fui-sidebar-sibling',
                template: `
                   <ng-content></ng-content>`,
                styles: [`
                   :host {
                       display: block;
                   }
               `]
            }] }
];
/** @nocollapse */
FuiSidebarSibling.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
FuiSidebarSibling.propDecorators = {
    isDimmedWhenVisible: [{ type: Input }],
    hasClasses: [{ type: HostBinding, args: ['class.pusher',] }],
    canCloseSidebar: [{ type: Input }],
    isVisible: [{ type: HostBinding, args: ['class.visible',] }],
    isDimmed: [{ type: HostBinding, args: ['class.dimmed',] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    FuiSidebarSibling.prototype.isDimmedWhenVisible;
    /** @type {?} */
    FuiSidebarSibling.prototype.hasClasses;
    /** @type {?} */
    FuiSidebarSibling.prototype.canCloseSidebar;
    /**
     * @type {?}
     * @private
     */
    FuiSidebarSibling.prototype._service;
    /**
     * @type {?}
     * @private
     */
    FuiSidebarSibling.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FuiSidebarSibling.prototype._element;
}
