// ngx-fomantic-ui/esm2015/modules/datepicker/classes/calendar-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CalendarMode } from '../services/calendar.service';
import { DateMappings, DatetimeMappings, MonthMappings, TimeMappings, YearMappings } from './calendar-mappings';
import { DatePrecision, DateUtil } from '../../../misc/util/internal';
/**
 * @abstract
 */
export class CalendarConfig {
    /**
     * @param {?} mode
     * @param {?} precision
     * @param {?} mappings
     * @param {?} fallback
     */
    constructor(mode, precision, mappings, fallback) {
        this.mode = mode;
        this.precision = precision;
        this.mappings = mappings;
        this.fallback = fallback;
    }
    /**
     * @param {?} providedDate
     * @return {?}
     */
    updateBounds(providedDate) {
        this.dateMinBound = DateUtil.startOf(DatePrecision.Year, new Date(), true);
        this.dateMinBound.setFullYear(0);
    }
}
if (false) {
    /** @type {?} */
    CalendarConfig.prototype.mode;
    /** @type {?} */
    CalendarConfig.prototype.precision;
    /** @type {?} */
    CalendarConfig.prototype.mappings;
    /** @type {?} */
    CalendarConfig.prototype.fallback;
    /** @type {?} */
    CalendarConfig.prototype.dateMinBound;
    /** @type {?} */
    CalendarConfig.prototype.dateMaxBound;
}
export class DateConfigBase extends CalendarConfig {
    /**
     * @param {?} precision
     * @param {?} mappings
     * @param {?} fallback
     */
    constructor(precision, mappings, fallback) {
        super(CalendarMode.DateOnly, precision, mappings, fallback);
    }
}
export class YearConfig extends DateConfigBase {
    constructor() {
        super(DatePrecision.Year, new YearMappings(), 'number');
    }
}
export class MonthConfig extends DateConfigBase {
    constructor() {
        super(DatePrecision.Month, new MonthMappings(), 'month');
    }
}
export class DateConfig extends DateConfigBase {
    constructor() {
        super(DatePrecision.Date, new DateMappings(), 'date');
    }
}
export class DatetimeConfig extends CalendarConfig {
    constructor() {
        super(CalendarMode.Both, DatePrecision.Minute, new DatetimeMappings(), 'datetime-local');
    }
}
export class TimeConfig extends CalendarConfig {
    constructor() {
        super(CalendarMode.TimeOnly, DatePrecision.Minute, new TimeMappings(), 'time');
    }
    /**
     * @param {?} providedDate
     * @return {?}
     */
    updateBounds(providedDate) {
        this.dateMaxBound = DateUtil.endOf(DatePrecision.Date, DateUtil.clone(providedDate));
        this.dateMinBound = DateUtil.previous(DatePrecision.Date, DateUtil.clone(this.dateMaxBound));
    }
}
