// ngx-fomantic-ui/esm2015/modules/datepicker/classes/date-parser.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DateFnsParser } from '../helpers/date-fns';
export class DateParser {
    /**
     * @param {?} format
     * @param {?} locale
     */
    constructor(format, locale) {
        this._format = format;
        this._parser = new DateFnsParser(locale);
    }
    /**
     * @param {?} date
     * @return {?}
     */
    format(date) {
        return this._parser.format(date, this._format);
    }
    /**
     * @param {?} dateString
     * @param {?=} baseDate
     * @return {?}
     */
    parse(dateString, baseDate = new Date()) {
        return this._parser.parse(dateString, this._format, baseDate);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    DateParser.prototype._format;
    /**
     * @type {?}
     * @private
     */
    DateParser.prototype._parser;
}
export class InternalDateParser extends DateParser {
    /**
     * @param {?} mode
     * @param {?} locale
     */
    constructor(mode, locale) {
        /** @type {?} */
        const internalFormats = {
            time: 'HH:mm',
            datetime: 'YYYY-MM-DDTHH:mm',
            date: 'YYYY-MM-DD',
            month: 'YYYY-MM',
            year: 'YYYY'
        };
        super(internalFormats[mode], locale);
    }
}
