// ngx-fomantic-ui/esm5/modules/select/directives/select-search.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, HostBinding, HostListener, Renderer2 } from '@angular/core';
var FuiSelectSearch = /** @class */ (function () {
    function FuiSelectSearch(_renderer, _element) {
        this._renderer = _renderer;
        this._element = _element;
        this.onQueryUpdated = new EventEmitter();
        this.onQueryKeyDown = new EventEmitter();
        this.hasClasses = true;
        this.autoComplete = 'off';
    }
    Object.defineProperty(FuiSelectSearch.prototype, "query", {
        set: /**
         * @param {?} query
         * @return {?}
         */
        function (query) {
            this._renderer.setProperty(this._element.nativeElement, 'value', query);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} query
     * @return {?}
     */
    FuiSelectSearch.prototype.updateQuery = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        this.onQueryUpdated.emit(query);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    FuiSelectSearch.prototype.onKeyDown = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.onQueryKeyDown.emit(e);
    };
    /**
     * @return {?}
     */
    FuiSelectSearch.prototype.focus = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Slightly delay to support in menu search.
        this._element.nativeElement.focus();
        setTimeout((/**
         * @return {?}
         */
        function () { return _this._element.nativeElement.focus(); }));
    };
    FuiSelectSearch.decorators = [
        { type: Directive, args: [{
                    selector: 'input[fuiSelectSearch]'
                },] }
    ];
    /** @nocollapse */
    FuiSelectSearch.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    FuiSelectSearch.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.search',] }],
        autoComplete: [{ type: HostBinding, args: ['attr.autocomplete',] }],
        updateQuery: [{ type: HostListener, args: ['input', ['$event.target.value'],] }],
        onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
    };
    return FuiSelectSearch;
}());
export { FuiSelectSearch };
if (false) {
    /** @type {?} */
    FuiSelectSearch.prototype.hasClasses;
    /** @type {?} */
    FuiSelectSearch.prototype.autoComplete;
    /** @type {?} */
    FuiSelectSearch.prototype.onQueryUpdated;
    /** @type {?} */
    FuiSelectSearch.prototype.onQueryKeyDown;
    /**
     * @type {?}
     * @private
     */
    FuiSelectSearch.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FuiSelectSearch.prototype._element;
}
