// ngx-fomantic-ui/esm2015/modules/progress/components/progress.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Input } from '@angular/core';
export class FuiProgress {
    constructor() {
        this.value = 0;
        this.maximum = 100;
        this.precision = 0;
        this._overrideSuccess = false;
        this.autoSuccess = true;
        this.showProgress = true;
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get reachedMaximum() {
        return this._overrideSuccess || ((this.value >= this.maximum) && this.autoSuccess);
    }
    /**
     * @return {?}
     */
    get percentage() {
        /** @type {?} */
        const boundedValue = Math.min(Math.max(this.value, 0), this.maximum);
        /** @type {?} */
        const percentage = (boundedValue / this.maximum) * 100;
        return percentage.toFixed(this.precision);
    }
    /**
     * @param {?} classes
     * @return {?}
     */
    set classValue(classes) {
        if (classes.includes('attached') || classes.includes('tiny')) {
            this.showProgress = false;
        }
        if (classes.includes('success')) {
            this._overrideSuccess = true;
        }
        this.transition = 'ease';
        this.transitionDuration = 350;
        this.canCompletelyEmpty = false;
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        // Convert value from string to number where necessary.
        /** @type {?} */
        const converted = +value;
        if (Number.isNaN(converted)) {
            return;
        }
        this._value = converted;
    }
    /**
     * @return {?}
     */
    get maximum() {
        return this._maximum;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maximum(value) {
        // Convert value from string to number where necessary.
        /** @type {?} */
        const converted = +value;
        if (Number.isNaN(converted)) {
            return;
        }
        this._maximum = converted;
    }
    /**
     * @return {?}
     */
    get precision() {
        return this._precision;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set precision(value) {
        // Convert value from string to number where necessary.
        /** @type {?} */
        const converted = +value;
        if (Number.isNaN(converted)) {
            return;
        }
        this._precision = Math.min(Math.max(converted, 0), 20);
    }
}
FuiProgress.decorators = [
    { type: Component, args: [{
                selector: 'fui-progress',
                template: `
<div class="bar"
    [style.width.%]="percentage"
    [style.minWidth]="canCompletelyEmpty ? 0 : null"
    [style.transitionTimingFunction]="transition"
    [style.transitionDuration.ms]="transitionDuration">
    <div class="progress" *ngIf="showProgress">{{ percentage }}%</div>
</div>
<div class="label">
    <ng-content></ng-content>
</div>
`,
                styles: [`
.bar {
    transition-duration: 300ms !important;
    z-index: 1;
}
`]
            }] }
];
/** @nocollapse */
FuiProgress.ctorParameters = () => [];
FuiProgress.propDecorators = {
    transition: [{ type: Input }],
    transitionDuration: [{ type: Input }],
    canCompletelyEmpty: [{ type: Input }],
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.progress',] }],
    autoSuccess: [{ type: Input }],
    showProgress: [{ type: Input }],
    reachedMaximum: [{ type: HostBinding, args: ['class.success',] }],
    percentage: [{ type: HostBinding, args: ['attr.data-percent',] }],
    classValue: [{ type: Input, args: ['class',] }],
    value: [{ type: Input }],
    maximum: [{ type: Input }],
    precision: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FuiProgress.prototype.transition;
    /** @type {?} */
    FuiProgress.prototype.transitionDuration;
    /** @type {?} */
    FuiProgress.prototype.canCompletelyEmpty;
    /** @type {?} */
    FuiProgress.prototype.hasClasses;
    /** @type {?} */
    FuiProgress.prototype.autoSuccess;
    /** @type {?} */
    FuiProgress.prototype.showProgress;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._overrideSuccess;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._value;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._maximum;
    /**
     * @type {?}
     * @private
     */
    FuiProgress.prototype._precision;
}
