// ngx-fomantic-ui/esm2015/modules/popup/classes/popup-template-controller.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { TemplateRef } from '@angular/core';
import { FuiPopupController } from './popup-controller';
import { PopupConfig } from './popup-config';
/** @type {?} */
const templateRef = TemplateRef;
/**
 * @record
 * @template T
 */
export function ITemplatePopupContext() { }
if (false) {
    /** @type {?|undefined} */
    ITemplatePopupContext.prototype.context;
}
/**
 * @record
 * @template T
 */
export function ITemplatePopupConfig() { }
if (false) {
    /** @type {?|undefined} */
    ITemplatePopupConfig.prototype.template;
    /** @type {?|undefined} */
    ITemplatePopupConfig.prototype.context;
}
/**
 * @template T
 */
export class TemplatePopupConfig extends PopupConfig {
}
if (false) {
    /** @type {?} */
    TemplatePopupConfig.prototype.template;
    /** @type {?} */
    TemplatePopupConfig.prototype.context;
}
/**
 * @template T
 */
export class FuiPopupTemplateController extends FuiPopupController {
    /**
     * @param {?} renderer
     * @param {?} element
     * @param {?} componentFactory
     * @param {?} config
     */
    constructor(renderer, element, componentFactory, config) {
        super(renderer, element, componentFactory, config);
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    configure(config) {
        super.configure(config);
        if (config) {
            this.template = config.template;
            this.context = config.context;
        }
    }
    /**
     * @return {?}
     */
    open() {
        // If there is a template, inject it into the view.
        if (this.template) {
            this.popup.templateSibling.clear();
            this._componentFactory.createView(this.popup.templateSibling, this.template, {
                $implicit: this.popup,
                context: this.context
            });
        }
        super.open();
    }
}
if (false) {
    /** @type {?} */
    FuiPopupTemplateController.prototype.template;
    /** @type {?} */
    FuiPopupTemplateController.prototype.context;
}
