// ngx-fomantic-ui/esm2015/modules/tabs/directives/tab-header.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, HostBinding, HostListener, Input, Output } from '@angular/core';
export class FuiTabHeader {
    constructor() {
        this._isActive = false;
        this.isActiveChange = new EventEmitter();
        this.isActiveExternalChange = new EventEmitter();
        this.onActivate = new EventEmitter();
        this.onDeactivate = new EventEmitter();
        this.isDisabled = false;
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get isActive() {
        return this._isActive;
    }
    /**
     * @param {?} active
     * @return {?}
     */
    set isActive(active) {
        /** @type {?} */
        let isActive = active;
        // Only used by @Input(), runs whenever user input changes `isActive`.
        // Run in timeout because `isDisabled` can prohibit user from changing `isActive`.
        // so update is delayed to avoid 'changed after checked' error.
        setTimeout((/**
         * @return {?}
         */
        () => {
            // Only allow change if tab header is not disabled.
            isActive = !this.isDisabled ? active : false;
            this.setActiveState(isActive);
            // Fire 'external change' event as user input has occured.
            this.isActiveExternalChange.emit(isActive);
        }));
    }
    /**
     * @return {?}
     */
    get isDisabled() {
        return this._isDisabled;
    }
    /**
     * @param {?} disabled
     * @return {?}
     */
    set isDisabled(disabled) {
        // Only update if value provided is different to current one.
        if (this._isDisabled !== disabled) {
            this._isDisabled = disabled;
            // If now disabled, then tab header must be deactivated.
            if (this.isDisabled) {
                this.isActive = false;
            }
        }
    }
    // Internally update active state.
    /**
     * @param {?} active
     * @return {?}
     */
    setActiveState(active) {
        // If (cast) active value has changed:
        if (!!this._isActive !== active) {
            // Update to the new value.
            this._isActive = active;
            // Fire the appropriate activation event.
            if (active) {
                this.onActivate.emit();
            }
            else {
                this.onDeactivate.emit();
            }
        }
        // Regardless, emit a change to `isActive`, so [(isActive)] works correctly.
        this.isActiveChange.emit(active);
    }
    /**
     * @return {?}
     */
    onClick() {
        if (!this.isDisabled) {
            // Activate the tab when clicked, so long as it isn't disabled.
            this.isActive = true;
        }
    }
}
FuiTabHeader.decorators = [
    { type: Directive, args: [{
                selector: '[fuiTabHeader]'
            },] }
];
/** @nocollapse */
FuiTabHeader.ctorParameters = () => [];
FuiTabHeader.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.item',] }],
    id: [{ type: Input, args: ['fuiTabHeader',] }],
    isActiveChange: [{ type: Output }],
    onActivate: [{ type: Output, args: ['activate',] }],
    onDeactivate: [{ type: Output, args: ['deactivate',] }],
    isActive: [{ type: HostBinding, args: ['class.active',] }, { type: Input }],
    isDisabled: [{ type: HostBinding, args: ['class.disabled',] }, { type: Input }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    FuiTabHeader.prototype.hasClasses;
    /** @type {?} */
    FuiTabHeader.prototype.id;
    /** @type {?} */
    FuiTabHeader.prototype.isActiveChange;
    /** @type {?} */
    FuiTabHeader.prototype.isActiveExternalChange;
    /** @type {?} */
    FuiTabHeader.prototype.onActivate;
    /** @type {?} */
    FuiTabHeader.prototype.onDeactivate;
    /**
     * @type {?}
     * @private
     */
    FuiTabHeader.prototype._isActive;
    /**
     * @type {?}
     * @private
     */
    FuiTabHeader.prototype._isDisabled;
}
