// ngx-fomantic-ui/esm5/modules/popup/popup.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FuiTransitionModule } from '../transition/internal';
import { FuiUtilityModule } from '../../misc/util/internal';
import { FuiPopupDirective } from './directives/popup.directive';
import { FuiPopupArrow } from './components/popup-arrow';
import { FuiPopup } from './components/popup';
import { FuiPopupConfig } from './services/popup.service';
var FuiPopupModule = /** @class */ (function () {
    function FuiPopupModule() {
    }
    FuiPopupModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FuiTransitionModule,
                        FuiUtilityModule
                    ],
                    declarations: [
                        FuiPopupDirective,
                        FuiPopupArrow,
                        FuiPopup
                    ],
                    exports: [
                        FuiPopupDirective,
                        FuiPopup
                    ],
                    providers: [
                        FuiPopupConfig
                    ],
                    entryComponents: [
                        FuiPopup
                    ]
                },] }
    ];
    return FuiPopupModule;
}());
export { FuiPopupModule };
