// ngx-fomantic-ui/esm2015/modules/accordion/services/accordion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class FuiAccordionService {
    constructor() {
        this.closeOthers = true;
        this.transition = 'fade';
        this.transitionDuration = 350;
        this.panels = [];
    }
    /**
     * @param {?} panel
     * @return {?}
     */
    addPanel(panel) {
        panel.service = this;
        this.panels.push(panel);
    }
    /**
     * @param {?} panel
     * @return {?}
     */
    closeOtherPanels(panel) {
        if (!this.closeOthers) {
            return;
        }
        this.panels.forEach((/**
         * @param {?} p
         * @return {?}
         */
        p => {
            if (p !== panel) {
                p.isOpen = false;
            }
        }));
    }
}
if (false) {
    /** @type {?} */
    FuiAccordionService.prototype.closeOthers;
    /** @type {?} */
    FuiAccordionService.prototype.transition;
    /** @type {?} */
    FuiAccordionService.prototype.transitionDuration;
    /** @type {?} */
    FuiAccordionService.prototype.panels;
}
