// ngx-fomantic-ui/esm5/modules/rating/components/rating.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Directive, EventEmitter, HostBinding, HostListener, Input, Output } from '@angular/core';
import { CustomValueAccessor, customValueAccessorFactory } from '../../../misc/util/internal';
var FuiRating = /** @class */ (function () {
    function FuiRating() {
        this.hoveredIndex = -1;
        this.value = 0;
        this.valueChange = new EventEmitter();
        this.type = 'star';
        this.maximum = 5;
        this.isReadonly = false;
        this.hasClasses = true;
    }
    Object.defineProperty(FuiRating.prototype, "icons", {
        get: /**
         * @return {?}
         */
        function () {
            // tslint:disable-next-line:prefer-literal
            return new Array(this.maximum);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiRating.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            return this._type;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._type = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiRating.prototype, "maximum", {
        get: /**
         * @return {?}
         */
        function () {
            return this._maximum;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._maximum = +value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} i
     * @return {?}
     */
    FuiRating.prototype.onClick = /**
     * @param {?} i
     * @return {?}
     */
    function (i) {
        if (!this.isReadonly) {
            this.value = i + 1;
            this.valueChange.emit(this.value);
        }
    };
    /**
     * @param {?} i
     * @return {?}
     */
    FuiRating.prototype.onMouseover = /**
     * @param {?} i
     * @return {?}
     */
    function (i) {
        this.hoveredIndex = i;
    };
    /**
     * @return {?}
     */
    FuiRating.prototype.onMouseout = /**
     * @return {?}
     */
    function () {
        this.hoveredIndex = -1;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FuiRating.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    FuiRating.decorators = [
        { type: Component, args: [{
                    selector: 'fui-rating',
                    // tslint:disable-next-line
                    template: "<i *ngFor=\"let icon of icons; let i = index\" class=\"icon {{type}}\" (mouseover)=\"onMouseover(i)\" (click)=\"onClick(i)\" [class.selected]=\"hoveredIndex >= i && !isReadonly\" [class.active]=\"value > i\"></i>",
                    styles: ["\n:host.read-only .icon {\n    cursor: auto\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiRating.ctorParameters = function () { return []; };
    FuiRating.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.rating',] }],
        valueChange: [{ type: Output }],
        isReadonly: [{ type: HostBinding, args: ['class.read-only',] }, { type: Input }],
        type: [{ type: Input }],
        maximum: [{ type: Input }],
        onMouseout: [{ type: HostListener, args: ['mouseout',] }]
    };
    return FuiRating;
}());
export { FuiRating };
if (false) {
    /** @type {?} */
    FuiRating.prototype.hasClasses;
    /** @type {?} */
    FuiRating.prototype.value;
    /** @type {?} */
    FuiRating.prototype.valueChange;
    /** @type {?} */
    FuiRating.prototype.isReadonly;
    /** @type {?} */
    FuiRating.prototype.hoveredIndex;
    /**
     * @type {?}
     * @private
     */
    FuiRating.prototype._type;
    /**
     * @type {?}
     * @private
     */
    FuiRating.prototype._maximum;
}
var FuiRatingValueAccessor = /** @class */ (function (_super) {
    tslib_1.__extends(FuiRatingValueAccessor, _super);
    function FuiRatingValueAccessor(host) {
        return _super.call(this, host) || this;
    }
    FuiRatingValueAccessor.decorators = [
        { type: Directive, args: [{
                    selector: 'fui-rating',
                    host: { '(valueChange)': 'onChange($event)' },
                    providers: [customValueAccessorFactory(FuiRatingValueAccessor)]
                },] }
    ];
    /** @nocollapse */
    FuiRatingValueAccessor.ctorParameters = function () { return [
        { type: FuiRating }
    ]; };
    return FuiRatingValueAccessor;
}(CustomValueAccessor));
export { FuiRatingValueAccessor };
