// ngx-fomantic-ui/esm5/modules/accordion/services/accordion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FuiAccordionService = /** @class */ (function () {
    function FuiAccordionService() {
        this.closeOthers = true;
        this.transition = 'fade';
        this.transitionDuration = 350;
        this.panels = [];
    }
    /**
     * @param {?} panel
     * @return {?}
     */
    FuiAccordionService.prototype.addPanel = /**
     * @param {?} panel
     * @return {?}
     */
    function (panel) {
        panel.service = this;
        this.panels.push(panel);
    };
    /**
     * @param {?} panel
     * @return {?}
     */
    FuiAccordionService.prototype.closeOtherPanels = /**
     * @param {?} panel
     * @return {?}
     */
    function (panel) {
        if (!this.closeOthers) {
            return;
        }
        this.panels.forEach((/**
         * @param {?} p
         * @return {?}
         */
        function (p) {
            if (p !== panel) {
                p.isOpen = false;
            }
        }));
    };
    return FuiAccordionService;
}());
export { FuiAccordionService };
if (false) {
    /** @type {?} */
    FuiAccordionService.prototype.closeOthers;
    /** @type {?} */
    FuiAccordionService.prototype.transition;
    /** @type {?} */
    FuiAccordionService.prototype.transitionDuration;
    /** @type {?} */
    FuiAccordionService.prototype.panels;
}
