// ngx-fomantic-ui/esm5/misc/util/helpers/custom-value-accessor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { forwardRef } from '@angular/core';
/**
 * @record
 * @template T
 */
export function ICustomValueAccessorHost() { }
if (false) {
    /**
     * @param {?} value
     * @return {?}
     */
    ICustomValueAccessorHost.prototype.writeValue = function (value) { };
}
/**
 * @template U, T
 */
var /**
 * @template U, T
 */
CustomValueAccessor = /** @class */ (function () {
    function CustomValueAccessor(_host) {
        this._host = _host;
        this.onChange = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
        });
        this.onTouched = (/**
         * @return {?}
         */
        function () {
        });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    CustomValueAccessor.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._host.writeValue(value);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CustomValueAccessor.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CustomValueAccessor.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    return CustomValueAccessor;
}());
/**
 * @template U, T
 */
export { CustomValueAccessor };
if (false) {
    /** @type {?} */
    CustomValueAccessor.prototype.onChange;
    /** @type {?} */
    CustomValueAccessor.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    CustomValueAccessor.prototype._host;
}
/**
 * @record
 */
export function IValueAccessorProvider() { }
if (false) {
    /** @type {?} */
    IValueAccessorProvider.prototype.provide;
    /** @type {?} */
    IValueAccessorProvider.prototype.useExisting;
    /** @type {?} */
    IValueAccessorProvider.prototype.multi;
}
/**
 * @param {?} type
 * @return {?}
 */
export function customValueAccessorFactory(type) {
    return {
        provide: NG_VALUE_ACCESSOR,
        useExisting: forwardRef((/**
         * @return {?}
         */
        function () { return type; })),
        multi: true
    };
}
