// ngx-fomantic-ui/esm5/misc/util/helpers/custom-validator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALIDATORS } from '@angular/forms';
import { forwardRef } from '@angular/core';
/**
 * @record
 */
export function ICustomValidatorHost() { }
if (false) {
    /**
     * @param {?} c
     * @return {?}
     */
    ICustomValidatorHost.prototype.validate = function (c) { };
}
/**
 * @template T
 */
var /**
 * @template T
 */
CustomValidator = /** @class */ (function () {
    function CustomValidator(_host) {
        this._host = _host;
        this.onValidatorChange = (/**
         * @return {?}
         */
        function () {
        });
    }
    /**
     * @param {?} c
     * @return {?}
     */
    CustomValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this._host.validate(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CustomValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onValidatorChange = fn;
    };
    return CustomValidator;
}());
/**
 * @template T
 */
export { CustomValidator };
if (false) {
    /** @type {?} */
    CustomValidator.prototype.onValidatorChange;
    /**
     * @type {?}
     * @private
     */
    CustomValidator.prototype._host;
}
/**
 * @record
 */
export function IValidationProvider() { }
if (false) {
    /** @type {?} */
    IValidationProvider.prototype.provide;
    /** @type {?} */
    IValidationProvider.prototype.useExisting;
    /** @type {?} */
    IValidationProvider.prototype.multi;
}
/**
 * @param {?} type
 * @return {?}
 */
export function customValidatorFactory(type) {
    return {
        provide: NG_VALIDATORS,
        useExisting: forwardRef((/**
         * @return {?}
         */
        function () { return type; })),
        multi: true
    };
}
