// ngx-fomantic-ui/esm2015/collections/message/classes/active-message.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class FuiActiveMessage {
}
if (false) {
    /**
     * @abstract
     * @param {?} callback
     * @return {?}
     */
    FuiActiveMessage.prototype.onClick = function (callback) { };
    /**
     * @abstract
     * @param {?} callback
     * @return {?}
     */
    FuiActiveMessage.prototype.onDismiss = function (callback) { };
    /**
     * @abstract
     * @return {?}
     */
    FuiActiveMessage.prototype.dismiss = function () { };
}
export class ActiveMessage {
    /**
     * @param {?} config
     * @param {?} componentRef
     */
    constructor(config, componentRef) {
        this.config = config;
        this.componentRef = componentRef;
        this.component.onDismiss.subscribe((/**
         * @return {?}
         */
        () => this.componentRef.destroy()));
    }
    /**
     * @return {?}
     */
    get component() {
        return this.componentRef.instance;
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    onClick(callback) {
        this.config.onClick.subscribe((/**
         * @return {?}
         */
        () => callback()));
        return this;
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    onDismiss(callback) {
        this.config.onDismiss.subscribe((/**
         * @return {?}
         */
        () => callback()));
        return this;
    }
    /**
     * @return {?}
     */
    dismiss() {
        this.component.dismiss();
    }
}
if (false) {
    /** @type {?} */
    ActiveMessage.prototype.config;
    /** @type {?} */
    ActiveMessage.prototype.componentRef;
}
