// ngx-fomantic-ui/esm2015/collections/message/classes/message-controller.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IMessageController() { }
if (false) {
    /** @type {?} */
    IMessageController.prototype.maxShown;
    /** @type {?} */
    IMessageController.prototype.isNewestOnTop;
    /**
     * @param {?} config
     * @return {?}
     */
    IMessageController.prototype.show = function (config) { };
    /**
     * @return {?}
     */
    IMessageController.prototype.dismissAll = function () { };
}
export class MessageController {
    constructor() {
        this.maxShown = 7;
        this.isNewestOnTop = true;
    }
    /**
     * @param {?} container
     * @return {?}
     */
    registerContainer(container) {
        this._container = container;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    show(config) {
        this.throwContainerError();
        return this._container.show(config, this.maxShown, this.isNewestOnTop);
    }
    /**
     * @return {?}
     */
    dismissAll() {
        this.throwContainerError();
        return this._container.dismissAll();
    }
    /**
     * @private
     * @return {?}
     */
    throwContainerError() {
        if (!this._container) {
            throw new Error('You must pass this controller to a message container.');
        }
    }
}
if (false) {
    /** @type {?} */
    MessageController.prototype.maxShown;
    /** @type {?} */
    MessageController.prototype.isNewestOnTop;
    /**
     * @type {?}
     * @private
     */
    MessageController.prototype._container;
}
