// ngx-fomantic-ui/esm2015/modules/modal/components/dimmer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, HostBinding, Renderer2 } from '@angular/core';
import { FuiDimmer } from '../../dimmer/internal';
export class FuiModalDimmer extends FuiDimmer {
    /**
     * @param {?} renderer
     * @param {?} element
     * @param {?} changeDetector
     */
    constructor(renderer, element, changeDetector) {
        super(renderer, element, changeDetector);
        this.hasClasses = true;
        this.isClickable = false;
    }
}
FuiModalDimmer.decorators = [
    { type: Component, args: [{
                selector: 'fui-modal-dimmer',
                template: `<ng-content></ng-content>`,
                styles: [`
        :host.ui.dimmer:not(.hidden) {
            transition: none;
            overflow-y: auto;
            display: flex !important;
        }
    `]
            }] }
];
/** @nocollapse */
FuiModalDimmer.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
FuiModalDimmer.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.page',] }, { type: HostBinding, args: ['class.modals',] }]
};
if (false) {
    /** @type {?} */
    FuiModalDimmer.prototype.hasClasses;
}
