// ngx-fomantic-ui/esm5/modules/transition/directives/transition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, HostBinding, Input, Renderer2 } from '@angular/core';
import { TransitionController } from '../classes/transition-controller';
var FuiTransition = /** @class */ (function () {
    function FuiTransition(_renderer, _element, _changeDetector) {
        this._renderer = _renderer;
        this._element = _element;
        this._changeDetector = _changeDetector;
        this.transitionClass = true;
    }
    Object.defineProperty(FuiTransition.prototype, "fuiTransition", {
        set: /**
         * @param {?} tC
         * @return {?}
         */
        function (tC) {
            // Set the transition controller (e.g. '<div [fuiTransition]="transitionController"></div>').
            this.setTransitionController(tC);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiTransition.prototype, "isVisible", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._controller) {
                return this._controller.isVisible;
            }
            return false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiTransition.prototype, "isHidden", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._controller) {
                return this._controller.isHidden;
            }
            return false;
        },
        enumerable: true,
        configurable: true
    });
    // Initialises the controller with the injected renderer and elementRef.
    // Initialises the controller with the injected renderer and elementRef.
    /**
     * @param {?} transitionController
     * @return {?}
     */
    FuiTransition.prototype.setTransitionController = 
    // Initialises the controller with the injected renderer and elementRef.
    /**
     * @param {?} transitionController
     * @return {?}
     */
    function (transitionController) {
        this._controller = transitionController;
        this._controller.registerRenderer(this._renderer);
        this._controller.registerElement(this._element.nativeElement);
        this._controller.registerChangeDetector(this._changeDetector);
    };
    FuiTransition.decorators = [
        { type: Directive, args: [{
                    selector: '[fuiTransition]',
                    exportAs: 'transition'
                },] }
    ];
    /** @nocollapse */
    FuiTransition.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    FuiTransition.propDecorators = {
        transitionClass: [{ type: HostBinding, args: ['class.transition',] }],
        fuiTransition: [{ type: Input }],
        isVisible: [{ type: HostBinding, args: ['class.visible',] }],
        isHidden: [{ type: HostBinding, args: ['class.hidden',] }]
    };
    return FuiTransition;
}());
export { FuiTransition };
if (false) {
    /** @type {?} */
    FuiTransition.prototype.transitionClass;
    /**
     * @type {?}
     * @private
     */
    FuiTransition.prototype._controller;
    /**
     * @type {?}
     * @protected
     */
    FuiTransition.prototype._renderer;
    /**
     * @type {?}
     * @protected
     */
    FuiTransition.prototype._element;
    /**
     * @type {?}
     * @private
     */
    FuiTransition.prototype._changeDetector;
}
