// ngx-fomantic-ui/esm5/modules/modal/modal.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FuiDimmerModule } from '../dimmer/internal';
import { FuiTransitionModule } from '../transition/internal';
import { FuiUtilityModule } from '../../misc/util/internal';
import { FuiModalService } from './services/modal.service';
import { FuiModal } from './components/modal';
import { FuiModalDimmer } from './components/dimmer';
var FuiModalModule = /** @class */ (function () {
    function FuiModalModule() {
    }
    FuiModalModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FuiDimmerModule,
                        FuiTransitionModule,
                        FuiUtilityModule
                    ],
                    declarations: [
                        FuiModal,
                        FuiModalDimmer
                    ],
                    exports: [
                        FuiModal
                    ],
                    providers: [
                        FuiModalService
                    ],
                    entryComponents: [
                        FuiModal
                    ]
                },] }
    ];
    return FuiModalModule;
}());
export { FuiModalModule };
