// ngx-fomantic-ui/esm5/modules/dimmer/components/dimmer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, Renderer2 } from '@angular/core';
import { FuiTransition, Transition, TransitionController, TransitionDirection } from '../../transition/internal';
var FuiDimmer = /** @class */ (function (_super) {
    tslib_1.__extends(FuiDimmer, _super);
    function FuiDimmer(renderer, element, changeDetector) {
        var _this = _super.call(this, renderer, element, changeDetector) || this;
        _this._isDimmed = false;
        _this.isDimmedChange = new EventEmitter();
        _this.isClickable = true;
        _this.wrapContent = true;
        _this.hasClasses = true;
        return _this;
    }
    Object.defineProperty(FuiDimmer.prototype, "isDimmed", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isDimmed;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var isDimmed = !!value;
            if (!this._transitionController) {
                // Initialise transition functionality when first setting dimmed, to ensure initial state doesn't transition.
                this._transitionController = new TransitionController(isDimmed, 'block');
                this.setTransitionController(this._transitionController);
                this._isDimmed = isDimmed;
            }
            else if (this._isDimmed !== isDimmed) {
                this._isDimmed = isDimmed;
                this._transitionController.stopAll();
                this._transitionController.animate(new Transition('fade', this.transitionDuration, isDimmed ? TransitionDirection.In : TransitionDirection.Out));
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FuiDimmer.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (this.isClickable) {
            this.isDimmed = false;
            this.isDimmedChange.emit(this.isDimmed);
        }
    };
    FuiDimmer.decorators = [
        { type: Component, args: [{
                    selector: 'fui-dimmer',
                    template: "\n<div [class.content]=\"wrapContent\">\n    <ng-content></ng-content>\n</div>\n",
                    styles: ["\n:host.dimmer:not(.hidden) {\n    transition: none;\n    display: flex !important;\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiDimmer.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    FuiDimmer.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.dimmer',] }],
        isDimmedChange: [{ type: Output }],
        isClickable: [{ type: Input }],
        transition: [{ type: Input }],
        transitionDuration: [{ type: Input }],
        wrapContent: [{ type: Input }],
        isDimmed: [{ type: HostBinding, args: ['class.active',] }, { type: Input }],
        onClick: [{ type: HostListener, args: ['click',] }]
    };
    return FuiDimmer;
}(FuiTransition));
export { FuiDimmer };
if (false) {
    /** @type {?} */
    FuiDimmer.prototype.hasClasses;
    /** @type {?} */
    FuiDimmer.prototype.isDimmedChange;
    /** @type {?} */
    FuiDimmer.prototype.isClickable;
    /** @type {?} */
    FuiDimmer.prototype.transition;
    /** @type {?} */
    FuiDimmer.prototype.transitionDuration;
    /** @type {?} */
    FuiDimmer.prototype.wrapContent;
    /**
     * @type {?}
     * @private
     */
    FuiDimmer.prototype._transitionController;
    /**
     * @type {?}
     * @private
     */
    FuiDimmer.prototype._isDimmed;
}
