// ngx-fomantic-ui/esm2015/modules/modal/classes/active-modal.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Helper class to support method chaining when calling `FuiModalService.open(...)`.
/**
 * @template T, U, V
 */
export class ActiveModal {
    /**
     * @param {?} instance
     * @param {?} componentRef
     */
    constructor(instance, componentRef) {
        this._config = instance;
        this._componentRef = componentRef;
        // Automatically destroy the modal component when it has been dismissed.
        this.component.onDismiss.subscribe((/**
         * @return {?}
         */
        () => this._componentRef.destroy()));
    }
    // Shorthand for direct access to the `FuiModal` instance.
    /**
     * @return {?}
     */
    get component() {
        return this._componentRef.instance;
    }
    // Registers a callback for when `onApprove` is fired.
    /**
     * @param {?} callback
     * @return {?}
     */
    onApprove(callback) {
        this.component.onApprove.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        (res) => callback(res)));
        return this;
    }
    // Registers a callback for when `onDeny` is fired.
    /**
     * @param {?} callback
     * @return {?}
     */
    onDeny(callback) {
        this.component.onDeny.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        (res) => callback(res)));
        return this;
    }
    // Fires the approve event of the modal manually.
    /**
     * @param {?} result
     * @return {?}
     */
    approve(result) {
        this.component.approve(result);
    }
    // Fires the deny event of the modal manually.
    /**
     * @param {?} result
     * @return {?}
     */
    deny(result) {
        this.component.deny(result);
    }
    // Removes the modal component instantly, without transitions or firing any events.
    /**
     * @return {?}
     */
    destroy() {
        this._componentRef.destroy();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActiveModal.prototype._config;
    /**
     * @type {?}
     * @private
     */
    ActiveModal.prototype._componentRef;
}
