// ngx-fomantic-ui/esm2015/modules/search/components/search-result.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Input, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { FuiComponentFactory } from '../../../misc/util/internal';
// See https://github.com/Microsoft/TypeScript/issues/13449.
/** @type {?} */
const templateRef = TemplateRef;
/**
 * @template T
 */
export class FuiSearchResult {
    /**
     * @param {?} componentFactory
     */
    constructor(componentFactory) {
        this.componentFactory = componentFactory;
        this.hasClasses = true;
        // By default we make this function return an empty string, for the brief moment when it isn't displaying the correct label.
        this.formatter = (/**
         * @param {?} value
         * @return {?}
         */
        value => '');
    }
    /**
     * @return {?}
     */
    get template() {
        return this._template;
    }
    /**
     * @param {?} template
     * @return {?}
     */
    set template(template) {
        this._template = template;
        if (this.template) {
            this.componentFactory.createView(this.templateSibling, this.template, {
                $implicit: this.value,
                query: this.query
            });
        }
    }
}
FuiSearchResult.decorators = [
    { type: Component, args: [{
                selector: 'fui-search-result',
                template: `
<span #templateSibling></span>
<span *ngIf="!template" [innerHTML]="formatter(value, query)"></span>
`
            }] }
];
/** @nocollapse */
FuiSearchResult.ctorParameters = () => [
    { type: FuiComponentFactory }
];
FuiSearchResult.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.result',] }],
    value: [{ type: Input }],
    query: [{ type: Input }],
    formatter: [{ type: Input }],
    templateSibling: [{ type: ViewChild, args: ['templateSibling', { read: ViewContainerRef, static: false },] }],
    template: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FuiSearchResult.prototype.hasClasses;
    /** @type {?} */
    FuiSearchResult.prototype.value;
    /** @type {?} */
    FuiSearchResult.prototype.query;
    /** @type {?} */
    FuiSearchResult.prototype.formatter;
    /** @type {?} */
    FuiSearchResult.prototype.templateSibling;
    /**
     * @type {?}
     * @private
     */
    FuiSearchResult.prototype._template;
    /** @type {?} */
    FuiSearchResult.prototype.componentFactory;
}
