// ngx-fomantic-ui/esm2015/modules/accordion/components/accordion.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, HostBinding, Input, QueryList } from '@angular/core';
import { FuiAccordionPanel } from './accordion-panel';
import { FuiAccordionService } from '../services/accordion.service';
export class FuiAccordion {
    constructor() {
        // Accordion service is unique to each set of panels.
        this._service = new FuiAccordionService();
        this.hasClasses = true;
    }
    /**
     * @return {?}
     */
    get closeOthers() {
        return this._service.closeOthers;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set closeOthers(value) {
        this._service.closeOthers = value;
    }
    /**
     * @param {?} transition
     * @return {?}
     */
    set transition(transition) {
        this._service.transition = transition;
    }
    /**
     * @param {?} duration
     * @return {?}
     */
    set transitionDuration(duration) {
        this._service.transitionDuration = duration;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.updatePanels();
        // Reconnect panels after they have updated.
        this._panels.changes.subscribe((/**
         * @return {?}
         */
        () => this.updatePanels()));
    }
    /**
     * @return {?}
     */
    updatePanels() {
        this._panels.forEach((/**
         * @param {?} p
         * @return {?}
         */
        p => this._service.addPanel(p)));
    }
}
FuiAccordion.decorators = [
    { type: Component, args: [{
                selector: 'fui-accordion',
                template: `
<ng-content></ng-content>
`,
                styles: [`
/* Fix for general styling issues */
:host {
    display: block;
}

/* Fix for styled border issue */
:host.styled fui-accordion-panel:first-child .title {
    border-top: none
}
`]
            }] }
];
/** @nocollapse */
FuiAccordion.ctorParameters = () => [];
FuiAccordion.propDecorators = {
    hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.accordion',] }],
    _panels: [{ type: ContentChildren, args: [FuiAccordionPanel,] }],
    closeOthers: [{ type: Input }],
    transition: [{ type: Input }],
    transitionDuration: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FuiAccordion.prototype.hasClasses;
    /**
     * @type {?}
     * @protected
     */
    FuiAccordion.prototype._service;
    /**
     * @type {?}
     * @protected
     */
    FuiAccordion.prototype._panels;
}
