// ngx-fomantic-ui/esm5/collections/message/components/message-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { ActiveMessage } from '../classes/active-message';
import { FuiMessage } from './message';
import { FuiComponentFactory } from '../../../misc/util/internal';
import { MessageController } from '../classes/message-controller';
var FuiMessageContainer = /** @class */ (function () {
    function FuiMessageContainer(_componentFactory, _element) {
        this._componentFactory = _componentFactory;
        this._element = _element;
        this._messages = [];
        this._queue = [];
    }
    Object.defineProperty(FuiMessageContainer.prototype, "controller", {
        set: /**
         * @param {?} controller
         * @return {?}
         */
        function (controller) {
            controller.registerContainer(this);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} config
     * @param {?} maxShown
     * @param {?} showNewestFirst
     * @return {?}
     */
    FuiMessageContainer.prototype.show = /**
     * @param {?} config
     * @param {?} maxShown
     * @param {?} showNewestFirst
     * @return {?}
     */
    function (config, maxShown, showNewestFirst) {
        var _this = this;
        /** @type {?} */
        var componentRef = this._componentFactory.createComponent(FuiMessage);
        componentRef.instance.loadConfig(config);
        /** @type {?} */
        var active = new ActiveMessage(config, componentRef)
            .onDismiss((/**
         * @return {?}
         */
        function () { return _this.onMessageClose(active, showNewestFirst); }));
        if (this._messages.length < maxShown) {
            this.open(active, showNewestFirst);
        }
        else {
            this.queue(active);
        }
        return active;
    };
    /**
     * @return {?}
     */
    FuiMessageContainer.prototype.dismissAll = /**
     * @return {?}
     */
    function () {
        this._queue = [];
        this._messages.forEach((/**
         * @param {?} m
         * @return {?}
         */
        function (m) { return m.dismiss(); }));
    };
    /**
     * @private
     * @param {?} message
     * @param {?} showNewestFirst
     * @return {?}
     */
    FuiMessageContainer.prototype.open = /**
     * @private
     * @param {?} message
     * @param {?} showNewestFirst
     * @return {?}
     */
    function (message, showNewestFirst) {
        this._messages.push(message);
        this._componentFactory.attachToView(message.componentRef, this.containerSibling);
        if (!showNewestFirst) {
            this._componentFactory.moveToElement(message.componentRef, this._element.nativeElement);
        }
        message.component.show();
    };
    /**
     * @private
     * @param {?} message
     * @return {?}
     */
    FuiMessageContainer.prototype.queue = /**
     * @private
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this._queue.push(message);
    };
    /**
     * @private
     * @param {?} message
     * @param {?} showNewestFirst
     * @return {?}
     */
    FuiMessageContainer.prototype.onMessageClose = /**
     * @private
     * @param {?} message
     * @param {?} showNewestFirst
     * @return {?}
     */
    function (message, showNewestFirst) {
        this._messages = this._messages.filter((/**
         * @param {?} m
         * @return {?}
         */
        function (m) { return m !== message; }));
        if (this._queue.length > 0) {
            var _a = tslib_1.__read(this._queue.slice(0, 1), 1), queued = _a[0];
            this.open(queued, showNewestFirst);
        }
    };
    FuiMessageContainer.decorators = [
        { type: Component, args: [{
                    selector: 'fui-message-container',
                    template: "\n<div #containerSibling></div>\n",
                    styles: ["\n:host {\n    display: block;\n}\n:host >>> fui-message {\n    display: block;\n    margin-bottom: 1rem;\n}\n:host >>> fui-message:last-of-type {\n    margin-bottom: 0;\n}\n:host >>> fui-message {\n    cursor: pointer;\n}\n"]
                }] }
    ];
    /** @nocollapse */
    FuiMessageContainer.ctorParameters = function () { return [
        { type: FuiComponentFactory },
        { type: ElementRef }
    ]; };
    FuiMessageContainer.propDecorators = {
        containerSibling: [{ type: ViewChild, args: ['containerSibling', { read: ViewContainerRef, static: false },] }],
        controller: [{ type: Input }]
    };
    return FuiMessageContainer;
}());
export { FuiMessageContainer };
if (false) {
    /** @type {?} */
    FuiMessageContainer.prototype.containerSibling;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._messages;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._queue;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._componentFactory;
    /**
     * @type {?}
     * @private
     */
    FuiMessageContainer.prototype._element;
}
