// ngx-fomantic-ui/esm5/collections/message/components/message-global-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostListener } from '@angular/core';
import { Util } from '../../../misc/util/internal';
/** @type {?} */
export var MessagePosition = {
    Top: (/** @type {?} */ ('top')),
    TopLeft: (/** @type {?} */ ('top-left')),
    TopRight: (/** @type {?} */ ('top-right')),
    Bottom: (/** @type {?} */ ('bottom')),
    BottomLeft: (/** @type {?} */ ('bottom-left')),
    BottomRight: (/** @type {?} */ ('bottom-right'))
};
var FuiMessageGlobalContainer = /** @class */ (function () {
    function FuiMessageGlobalContainer() {
    }
    Object.defineProperty(FuiMessageGlobalContainer.prototype, "dynamicClasses", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var classes = {};
            this.position
                .split('-')
                .forEach((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return classes[p] = true; }));
            return classes;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FuiMessageGlobalContainer.prototype, "dynamicWidth", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var margin = Util.DOM.getDocumentFontSize();
            /** @type {?} */
            var width = this.width;
            if (this.position === MessagePosition.Top ||
                this.position === MessagePosition.Bottom ||
                window.innerWidth < width + margin * 2) {
                width = window.innerWidth - margin * 2;
            }
            return width;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FuiMessageGlobalContainer.prototype.onDocumentResize = /**
     * @return {?}
     */
    function () {
    };
    FuiMessageGlobalContainer.decorators = [
        { type: Component, args: [{
                    selector: 'fui-message-global-container',
                    template: "\n<div class=\"global container\" [ngClass]=\"dynamicClasses\" [style.width.px]=\"dynamicWidth\">\n    <fui-message-container [controller]=\"controller\"></fui-message-container>\n</div>\n",
                    styles: ["\n.global.container {\n    display: block;\n    position: fixed;\n}\n.global.container.top {\n    top: 1rem;\n}\n.global.container.bottom {\n    bottom: 1rem;\n}\n.global.container.left {\n    left: 1rem;\n}\n.global.container.right {\n    right: 1rem;\n}\n.global.container:not(.left):not(.right) {\n    left: 1rem;\n}\n"]
                }] }
    ];
    FuiMessageGlobalContainer.propDecorators = {
        onDocumentResize: [{ type: HostListener, args: ['window:resize',] }]
    };
    return FuiMessageGlobalContainer;
}());
export { FuiMessageGlobalContainer };
if (false) {
    /** @type {?} */
    FuiMessageGlobalContainer.prototype.controller;
    /** @type {?} */
    FuiMessageGlobalContainer.prototype.position;
    /** @type {?} */
    FuiMessageGlobalContainer.prototype.width;
}
