// ngx-fomantic-ui/esm5/modules/select/components/multi-select-label.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, Renderer2, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { FuiTransition, Transition, TransitionController, TransitionDirection } from '../../transition/internal';
import { HandledEvent, FuiComponentFactory } from '../../../misc/util/internal';
// See https://github.com/Microsoft/TypeScript/issues/13449.
/** @type {?} */
var templateRef = TemplateRef;
/**
 * @template T
 */
var FuiMultiSelectLabel = /** @class */ (function (_super) {
    tslib_1.__extends(FuiMultiSelectLabel, _super);
    function FuiMultiSelectLabel(renderer, element, changeDetector, componentFactory) {
        var _this = _super.call(this, renderer, element, changeDetector) || this;
        _this.componentFactory = componentFactory;
        // Initialise transition functionality.
        _this._transitionController = new TransitionController(false, 'inline-block');
        _this.setTransitionController(_this._transitionController);
        _this.onDeselected = new EventEmitter();
        _this.hasClasses = true;
        _this._transitionController.animate(new Transition('scale', 100, TransitionDirection.In));
        return _this;
    }
    Object.defineProperty(FuiMultiSelectLabel.prototype, "template", {
        get: /**
         * @return {?}
         */
        function () {
            return this._template;
        },
        set: /**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            this._template = template;
            if (this.template) {
                this.componentFactory.createView(this.templateSibling, this.template, {
                    $implicit: this.value,
                    query: this.query
                });
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} e
     * @return {?}
     */
    FuiMultiSelectLabel.prototype.deselectOption = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        e.eventHandled = true;
        this._transitionController.animate(new Transition('scale', 100, TransitionDirection.Out, (/**
         * @return {?}
         */
        function () {
            return _this.onDeselected.emit(_this.value);
        })));
    };
    /**
     * @param {?} e
     * @return {?}
     */
    FuiMultiSelectLabel.prototype.onClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.eventHandled = true;
    };
    FuiMultiSelectLabel.decorators = [
        { type: Component, args: [{
                    selector: 'fui-multi-select-label',
                    template: "\n<span #templateSibling></span>\n<span *ngIf=\"!template\" [innerHTML]=\"formatter(value)\"></span>\n<i class=\"delete icon\" (click)=\"deselectOption($event)\"></i>\n"
                }] }
    ];
    /** @nocollapse */
    FuiMultiSelectLabel.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: FuiComponentFactory }
    ]; };
    FuiMultiSelectLabel.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.ui',] }, { type: HostBinding, args: ['class.label',] }],
        value: [{ type: Input }],
        query: [{ type: Input }],
        onDeselected: [{ type: Output, args: ['deselected',] }],
        formatter: [{ type: Input }],
        templateSibling: [{ type: ViewChild, args: ['templateSibling', { read: ViewContainerRef, static: false },] }],
        template: [{ type: Input }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return FuiMultiSelectLabel;
}(FuiTransition));
export { FuiMultiSelectLabel };
if (false) {
    /** @type {?} */
    FuiMultiSelectLabel.prototype.hasClasses;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.value;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.query;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.onDeselected;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.formatter;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.templateSibling;
    /**
     * @type {?}
     * @private
     */
    FuiMultiSelectLabel.prototype._transitionController;
    /**
     * @type {?}
     * @private
     */
    FuiMultiSelectLabel.prototype._template;
    /** @type {?} */
    FuiMultiSelectLabel.prototype.componentFactory;
}
