// ngx-fomantic-ui/esm5/modules/select/components/select-option.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, Renderer2, ViewChild, ViewContainerRef } from '@angular/core';
import { FuiDropdownMenuItem } from '../../dropdown/internal';
import { HandledEvent } from '../../../misc/util/internal';
/**
 * @template T
 */
var FuiSelectOption = /** @class */ (function (_super) {
    tslib_1.__extends(FuiSelectOption, _super);
    function FuiSelectOption(renderer, element, changeDetector) {
        var _this = 
        // We inherit FuiDropdownMenuItem to automatically gain all keyboard navigation functionality.
        // This is not done via adding the .item class because it isn't supported by Angular.
        _super.call(this, renderer, element) || this;
        _this.changeDetector = changeDetector;
        _this.hasClasses = true;
        _this.isActive = false;
        _this.onSelected = new EventEmitter();
        // By default we make the default text an empty label, for the brief moment when it isn't displaying the correct one.
        _this.renderedText = '';
        _this.usesTemplate = false;
        return _this;
    }
    Object.defineProperty(FuiSelectOption.prototype, "formatter", {
        set: /**
         * @param {?} formatter
         * @return {?}
         */
        function (formatter) {
            if (!this.usesTemplate) {
                this.renderedText = formatter(this.value);
            }
            else {
                this.renderedText = '';
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} e
     * @return {?}
     */
    FuiSelectOption.prototype.onClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        e.eventHandled = true;
        setTimeout((/**
         * @return {?}
         */
        function () { return _this.onSelected.emit(_this.value); }));
    };
    FuiSelectOption.decorators = [
        { type: Component, args: [{
                    selector: 'fui-select-option',
                    template: "\n<span #templateSibling></span>\n<span [innerHTML]=\"renderedText\"></span>\n"
                }] }
    ];
    /** @nocollapse */
    FuiSelectOption.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    FuiSelectOption.propDecorators = {
        hasClasses: [{ type: HostBinding, args: ['class.item',] }],
        value: [{ type: Input }],
        onSelected: [{ type: Output }],
        isActive: [{ type: HostBinding, args: ['class.active',] }],
        templateSibling: [{ type: ViewChild, args: ['templateSibling', { read: ViewContainerRef, static: false },] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return FuiSelectOption;
}(FuiDropdownMenuItem));
export { FuiSelectOption };
if (false) {
    /** @type {?} */
    FuiSelectOption.prototype.hasClasses;
    /** @type {?} */
    FuiSelectOption.prototype.value;
    /** @type {?} */
    FuiSelectOption.prototype.onSelected;
    /** @type {?} */
    FuiSelectOption.prototype.isActive;
    /** @type {?} */
    FuiSelectOption.prototype.renderedText;
    /** @type {?} */
    FuiSelectOption.prototype.usesTemplate;
    /** @type {?} */
    FuiSelectOption.prototype.templateSibling;
    /** @type {?} */
    FuiSelectOption.prototype.changeDetector;
}
